// ht-client/dist/es/entities/actions/actions-list-client.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { EntityListClient } from "../../base/list-client";
import * as fromRoot from "../../reducers";
import { PageResults$ } from "ht-data";
import * as fromAction from "../../dispatchers/actions-dispatcher";
import { listQueryMixin } from "../../mixins/entity-query";
import { getFirstDataMixin } from "../../mixins/get-first-data";
import { clientSubMixin } from "../../mixins/client-subscription";
import { getPageDataMixin } from "../../mixins/get-data";
var ActionsList = (function (_super) {
    __extends(ActionsList, _super);
    function ActionsList(_a) {
        var dateRange = _a.dateRange, store = _a.store, dateParam = _a.dateParam, api = _a.api;
        var _this = _super.call(this) || this;
        _this.api$ = function (query) {
            return api.index(query);
        };
        _this.dateRange = dateRange;
        _this.store = store;
        _this.query$ = _this.store.select(fromRoot.getActionsListQuery);
        _this.dateParam = dateParam;
        _this.active$ = _this.store.select(fromRoot.getActionsListActive);
        _this.data$ = _this.store.select(fromRoot.getActionsList);
        _this.loading$ = _this.store.select(fromRoot.getActionsListLoading);
        _this.dataArray$ = _this.data$.pipe(PageResults$);
        return _this;
        // this.init()
    }
    ActionsList.prototype.setActive = function (active) {
        if (active === void 0) { active = true; }
        this.store.dispatch(new fromAction.SetListActive(active));
    };
    ActionsList.prototype.setLoading = function (loading) {
        this.store.dispatch(new fromAction.SetListLoading(loading));
    };
    ;
    ActionsList.prototype.setData = function (data) {
        this.store.dispatch(new fromAction.SetList(data));
    };
    ;
    ActionsList.prototype.setQuery = function (query) {
        this.store.dispatch(new fromAction.SetListQuery(query));
    };
    return ActionsList;
}(EntityListClient));
export { ActionsList };
;
export var ActionsListClient = clientSubMixin(getPageDataMixin(getFirstDataMixin(listQueryMixin(ActionsList))));
