// time-aware-polyline/dist/es/client/src/dispatchers/groups-dispatcher.js
export var SET_LIST_ACTIVE = "[GROUPS] set list active";
export var SET_ID = "[GROUPS] set id";
export var SET_GROUPS = "[GROUPS] set groups";
export var SET_GROUP = "[GROUPS] set group";
var SetListActive = (function () {
    function SetListActive(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = SET_LIST_ACTIVE;
    }
    return SetListActive;
}());
export { SetListActive };
var SetGroupId = (function () {
    function SetGroupId(payload) {
        this.payload = payload;
        this.type = SET_ID;
    }
    return SetGroupId;
}());
export { SetGroupId };
var SetGroupsAll = (function () {
    function SetGroupsAll(payload) {
        this.payload = payload;
        this.type = SET_GROUPS;
    }
    return SetGroupsAll;
}());
export { SetGroupsAll };
var SetGroup = (function () {
    function SetGroup(payload) {
        this.payload = payload;
        this.type = SET_GROUP;
    }
    return SetGroup;
}());
export { SetGroup };
