// time-aware-polyline/dist/es/client/src/dispatchers/query-dispatcher.js
export var SET_DATE_RANGE = "[QUERY] set date range";
export var SET_USER_QUERY = "[QUERY] set user query";
export var SET_USER_QUERY_RESET_PAGE = "[QUERY] set user query reset page";
export var CLEAR_USER_QUERY_KEY = "[QUERY] clear user query key";
export var SET_USER_ID = "[USERS] set user id";
export var TOGGLE_USER_ID = "[USERS] toggle user id";
export var SET_PLACELINE_QUERY = "[USERS] Set placeline query";
var SetDateRange = (function () {
    function SetDateRange(payload) {
        this.payload = payload;
        this.type = SET_DATE_RANGE;
    }
    return SetDateRange;
}());
export { SetDateRange };
var SetUserQuery = (function () {
    function SetUserQuery(payload) {
        this.payload = payload;
        this.type = SET_USER_QUERY;
    }
    return SetUserQuery;
}());
export { SetUserQuery };
var SetUserQueryResetPage = (function () {
    function SetUserQueryResetPage(payload) {
        this.payload = payload;
        this.type = SET_USER_QUERY_RESET_PAGE;
    }
    return SetUserQueryResetPage;
}());
export { SetUserQueryResetPage };
var ClearUserQueryKey = (function () {
    function ClearUserQueryKey(payload) {
        this.payload = payload;
        this.type = CLEAR_USER_QUERY_KEY;
    }
    return ClearUserQueryKey;
}());
export { ClearUserQueryKey };
var SetUserId = (function () {
    function SetUserId(payload) {
        this.payload = payload;
        this.type = SET_USER_ID;
    }
    return SetUserId;
}());
export { SetUserId };
var ToggleUserId = (function () {
    function ToggleUserId(payload) {
        this.payload = payload;
        this.type = TOGGLE_USER_ID;
    }
    return ToggleUserId;
}());
export { ToggleUserId };
var SetPlacelineQuery = (function () {
    function SetPlacelineQuery(payload) {
        if (payload === void 0) { payload = {}; }
        this.payload = payload;
        this.type = SET_PLACELINE_QUERY;
    }
    return SetPlacelineQuery;
}());
export { SetPlacelineQuery };
