// time-aware-polyline/dist/es/client/src/dispatchers/segments-dispatcher.js
export var SET_SELECTED_ID = "[SEGMENTS] set selected id";
export var SET_HIGHLIGHTED_ID = "[SEGMENTS] set highlighted id";
export var SET_RESET_MAP_ID = "[SEGMENTS] set reset map id";
export var SET_POPUP_ID = "[SEGMENTS] set popup id";
var SetSelectedId = (function () {
    function SetSelectedId(payload) {
        this.payload = payload;
        this.type = SET_SELECTED_ID;
    }
    return SetSelectedId;
}());
export { SetSelectedId };
var SetHighlightedId = (function () {
    function SetHighlightedId(payload) {
        this.payload = payload;
        this.type = SET_HIGHLIGHTED_ID;
    }
    return SetHighlightedId;
}());
export { SetHighlightedId };
var SetResetMapId = (function () {
    function SetResetMapId(payload) {
        this.payload = payload;
        this.type = SET_RESET_MAP_ID;
    }
    return SetResetMapId;
}());
export { SetResetMapId };
var SetPopupId = (function () {
    function SetPopupId(payload) {
        this.payload = payload;
        this.type = SET_POPUP_ID;
    }
    return SetPopupId;
}());
export { SetPopupId };
