// time-aware-polyline/dist/es/client/src/dispatchers/user-dispatcher.js
//placeline
export var SET_USER_DATA = "[USERS] set user data";
export var SET_PLACELINE_ID = "[USERS] set placeline id";
export var TOGGLE_PLACELINE_ID = "[USERS] toggle placeline id";
export var SET_PLACELINE_QUERY = "[USERS] set placeline query";
export var ADD_PLACELINE_QUERY = "[USERS] add placeline query";
export var SET_PLACELINE_LOADING = "[USERS] set placeline loading";
//list
export var SET_LIST_ACTIVE = "[USERS] set list active";
export var SET_LIST_ID = "[USERS] set list id";
export var SET_LIST_QUERY = "[USERS] set list query";
export var ADD_LIST_QUERY = "[USERS] add list query";
export var CLEAR_QUERY_KEY = "[USERS] clear query key";
export var TOGGLE_LIST_ID = "[USERS] toggle list id";
//analytics page
export var SET_USERS_ANALYTICS_PAGE = "[USERS] set user analytics page";
export var SET_USERS_ANALYTICS_ID = "[USERS] set analytics id";
export var SET_USERS_ANALYTICS_QUERY = "[USERS] set analytics query";
export var SET_USERS_ANALYTICS_LOADING = "[USERS] set analytics loading";
export var SET_USERS_LIST_DATA_MAP = "[USERS] set users list data map";
//analytics all
export var SET_MARKERS_ACTIVE = "[USERS] set markerS active";
export var SET_USERS_ANALYTICS_ALL = "[USERS] set users analytics all";
export var ADD_USERS_ANALYTICS_ALL = "[USERS] add users analytics all";
export var SET_USERS_ANALYTICS_ALL_QUERY = "[USERS] add analytics all query";
export var ADD_USERS_ANALYTICS_ALL_QUERY = "[USERS] set analytics all query";
export var SET_USERS_ANALYTICS_ALL_LOADING = "[USERS] set analytics all loading";
export var SET_USERS_MARKERS_DATA_MAP = "[USERS] set users listAll data map";
//summary
export var SET_SUMMARY_ACTIVE = "[USERS] set summary active";
export var SET_USERS_SUMMARY = "[USERS] set users summary";
export var SET_USERS_SUMMARY_LOADING = "[USERS] set summary loading";
export var SET_USERS_INDEX_PAGE = "[USERS] set user index page";
export var ADD_USERS_INDEX_ALL = "[USERS] add users index all";
export var SET_USERS_LIST_API_TYPE = "[USERS] set users list api type";
var SetUserData = (function () {
    function SetUserData(payload) {
        this.payload = payload;
        this.type = SET_USER_DATA;
    }
    return SetUserData;
}());
export { SetUserData };
var SetPlacelineId = (function () {
    function SetPlacelineId(payload) {
        this.payload = payload;
        this.type = SET_PLACELINE_ID;
    }
    return SetPlacelineId;
}());
export { SetPlacelineId };
var TogglePlacelineId = (function () {
    function TogglePlacelineId(payload) {
        this.payload = payload;
        this.type = TOGGLE_PLACELINE_ID;
    }
    return TogglePlacelineId;
}());
export { TogglePlacelineId };
var SetPlacelineQuery = (function () {
    function SetPlacelineQuery(payload) {
        this.payload = payload;
        this.type = SET_PLACELINE_QUERY;
    }
    return SetPlacelineQuery;
}());
export { SetPlacelineQuery };
var AddPlacelineQuery = (function () {
    function AddPlacelineQuery(payload) {
        this.payload = payload;
        this.type = ADD_PLACELINE_QUERY;
    }
    return AddPlacelineQuery;
}());
export { AddPlacelineQuery };
var SetPlacelineLoading = (function () {
    function SetPlacelineLoading(payload) {
        this.payload = payload;
        this.type = SET_PLACELINE_LOADING;
    }
    return SetPlacelineLoading;
}());
export { SetPlacelineLoading };
var SetUsersListId = (function () {
    function SetUsersListId(payload) {
        this.payload = payload;
        this.type = SET_LIST_ID;
    }
    return SetUsersListId;
}());
export { SetUsersListId };
var ToggleUsersListId = (function () {
    function ToggleUsersListId(payload) {
        this.payload = payload;
        this.type = TOGGLE_LIST_ID;
    }
    return ToggleUsersListId;
}());
export { ToggleUsersListId };
var SetListQuery = (function () {
    function SetListQuery(payload) {
        this.payload = payload;
        this.type = SET_LIST_QUERY;
    }
    return SetListQuery;
}());
export { SetListQuery };
var AddListQuery = (function () {
    function AddListQuery(payload) {
        this.payload = payload;
        this.type = ADD_LIST_QUERY;
    }
    return AddListQuery;
}());
export { AddListQuery };
var ClearQueryKey = (function () {
    function ClearQueryKey(payload) {
        this.payload = payload;
        this.type = CLEAR_QUERY_KEY;
    }
    return ClearQueryKey;
}());
export { ClearQueryKey };
var SetUsersAnalyticsPage = (function () {
    function SetUsersAnalyticsPage(payload) {
        this.payload = payload;
        this.type = SET_USERS_ANALYTICS_PAGE;
    }
    return SetUsersAnalyticsPage;
}());
export { SetUsersAnalyticsPage };
var SetUsersAnalyticsLoading = (function () {
    function SetUsersAnalyticsLoading(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = SET_USERS_ANALYTICS_LOADING;
    }
    return SetUsersAnalyticsLoading;
}());
export { SetUsersAnalyticsLoading };
var SetUsersIndexPage = (function () {
    function SetUsersIndexPage(payload) {
        this.payload = payload;
        this.type = SET_USERS_INDEX_PAGE;
    }
    return SetUsersIndexPage;
}());
export { SetUsersIndexPage };
var AddUsersAnalyticsAll = (function () {
    function AddUsersAnalyticsAll(payload) {
        this.payload = payload;
        this.type = ADD_USERS_ANALYTICS_ALL;
    }
    return AddUsersAnalyticsAll;
}());
export { AddUsersAnalyticsAll };
var SetUsersAnalyticsAll = (function () {
    function SetUsersAnalyticsAll(payload) {
        this.payload = payload;
        this.type = SET_USERS_ANALYTICS_ALL;
    }
    return SetUsersAnalyticsAll;
}());
export { SetUsersAnalyticsAll };
var SetUsersAnalyticsAllLoading = (function () {
    function SetUsersAnalyticsAllLoading(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = SET_USERS_ANALYTICS_ALL_LOADING;
    }
    return SetUsersAnalyticsAllLoading;
}());
export { SetUsersAnalyticsAllLoading };
var SetUsersIndexAll = (function () {
    function SetUsersIndexAll(payload) {
        this.payload = payload;
        //todo change name
        this.type = ADD_USERS_INDEX_ALL;
    }
    return SetUsersIndexAll;
}());
export { SetUsersIndexAll };
var SetUsersListApiType = (function () {
    function SetUsersListApiType(payload) {
        this.payload = payload;
        this.type = SET_USERS_LIST_API_TYPE;
    }
    return SetUsersListApiType;
}());
export { SetUsersListApiType };
var SetUsersListDataMap = (function () {
    function SetUsersListDataMap(payload) {
        this.payload = payload;
        this.type = SET_USERS_LIST_DATA_MAP;
    }
    return SetUsersListDataMap;
}());
export { SetUsersListDataMap };
var SetUsersMarkersDataMap = (function () {
    function SetUsersMarkersDataMap(payload) {
        this.payload = payload;
        this.type = SET_USERS_MARKERS_DATA_MAP;
    }
    return SetUsersMarkersDataMap;
}());
export { SetUsersMarkersDataMap };
var SetListActive = (function () {
    function SetListActive(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = SET_LIST_ACTIVE;
    }
    return SetListActive;
}());
export { SetListActive };
var SetMarkersActive = (function () {
    function SetMarkersActive(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = SET_MARKERS_ACTIVE;
    }
    return SetMarkersActive;
}());
export { SetMarkersActive };
var SetSummaryActive = (function () {
    function SetSummaryActive(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = SET_SUMMARY_ACTIVE;
    }
    return SetSummaryActive;
}());
export { SetSummaryActive };
var SetUsersSummary = (function () {
    function SetUsersSummary(payload) {
        this.payload = payload;
        this.type = SET_USERS_SUMMARY;
    }
    return SetUsersSummary;
}());
export { SetUsersSummary };
var SetSummaryLoading = (function () {
    function SetSummaryLoading(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = SET_USERS_SUMMARY_LOADING;
    }
    return SetSummaryLoading;
}());
export { SetSummaryLoading };
