// ht-client/dist/es/dispatchers/accounts-dispatcher.js
export var SET_ACCOUNT_USER = "[ACCOUNT] set account user";
export var SET_MEMBERHSHIPS_ALL = "[ACCOUNT] set memberships all";
export var SET_KEY = "[ACCOUNT] set key";
export var SET_TEMP_KEY = "[ACCOUNT] set temp key";
export var SET_USER_ID = "[ACCOUNT] set user id";
var SetAccountUser = (function () {
    function SetAccountUser(payload) {
        this.payload = payload;
        this.type = SET_ACCOUNT_USER;
    }
    return SetAccountUser;
}());
export { SetAccountUser };
var SetMembershipsAll = (function () {
    function SetMembershipsAll(payload) {
        this.payload = payload;
        this.type = SET_MEMBERHSHIPS_ALL;
    }
    return SetMembershipsAll;
}());
export { SetMembershipsAll };
var SetKey = (function () {
    function SetKey(payload) {
        this.payload = payload;
        this.type = SET_KEY;
    }
    return SetKey;
}());
export { SetKey };
var SetTempKey = (function () {
    function SetTempKey(payload) {
        this.payload = payload;
        this.type = SET_TEMP_KEY;
    }
    return SetTempKey;
}());
export { SetTempKey };
var SetUserId = (function () {
    function SetUserId(payload) {
        this.payload = payload;
        this.type = SET_USER_ID;
    }
    return SetUserId;
}());
export { SetUserId };
