// ht-client/dist/es/dispatchers/actions-dispatcher.js
/*
List
 */
export var SET_ACTIONS_LIST = "[ACTIONS] set actions list";
export var SET_ACTIONS_LIST_QUERY = "[ACTIONS] set actions list query";
export var ADD_ACTIONS_LIST_QUERY = "[ACTIONS] add actions list query";
export var SET_ACTIONS_LIST_ACTIVE = "[ACTIONS] set actions list active";
export var SET_ACTIONS_LIST_LOADING = "[ACTIONS] set actions list loading";
/*
Graph
 */
export var SET_ACTIONS_GRAPH = "[ACTIONS] set actions graph";
export var SET_ACTIONS_GRAPH_QUERY = "[ACTIONS] set actions graph query";
export var ADD_ACTIONS_GRAPH_QUERY = "[ACTIONS] add actions graph query";
export var SET_ACTIONS_GRAPH_LOADING = "[ACTIONS] set actions graph loading";
/*
Summary
 */
export var SET_ACTIONS_SUMMARY = "[ACTIONS] set actions summary";
export var SET_ACTIONS_SUMMARY_ACTIVE = "[ACTIONS] set actions summary active";
export var ADD_ACTIONS_SUMMARY_QUERY = "[ACTIONS] actions summary query";
export var SET_ACTIONS_SUMMARY_QUERY = "[ACTIONS] add set actions summary query";
export var SET_ACTIONS_SUMMARY_LOADING = "[ACTIONS] set actions summary loading";
/*
Heatmap
 */
// export const SET_ACTIONS_HEATMAP = "[ACTIONS] set actions summary";
// export const SET_ACTIONS_HEATMAP_ACTIVE = "[ACTIONS] set actions summary active";
// export const ADD_ACTIONS_HEATMAP_QUERY = "[ACTIONS] actions summary query";
// export const SET_ACTIONS_HEATMAP_QUERY = "[ACTIONS] add set actions summary query";
// export const SET_ACTIONS_HEATMAP_LOADING = "[ACTIONS] set actions graph loading";
var SetList = (function () {
    function SetList(payload) {
        this.payload = payload;
        this.type = SET_ACTIONS_LIST;
    }
    return SetList;
}());
export { SetList };
var SetListQuery = (function () {
    function SetListQuery(payload) {
        this.payload = payload;
        this.type = SET_ACTIONS_LIST_QUERY;
    }
    return SetListQuery;
}());
export { SetListQuery };
var AddListQuery = (function () {
    function AddListQuery(payload) {
        this.payload = payload;
        this.type = ADD_ACTIONS_LIST_QUERY;
    }
    return AddListQuery;
}());
export { AddListQuery };
var SetListActive = (function () {
    function SetListActive(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = SET_ACTIONS_LIST_ACTIVE;
    }
    return SetListActive;
}());
export { SetListActive };
var SetListLoading = (function () {
    function SetListLoading(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = SET_ACTIONS_LIST_LOADING;
    }
    return SetListLoading;
}());
export { SetListLoading };
var SetGraph = (function () {
    function SetGraph(payload) {
        this.payload = payload;
        this.type = SET_ACTIONS_GRAPH;
    }
    return SetGraph;
}());
export { SetGraph };
var SetGraphQuery = (function () {
    function SetGraphQuery(payload) {
        this.payload = payload;
        this.type = SET_ACTIONS_GRAPH_QUERY;
    }
    return SetGraphQuery;
}());
export { SetGraphQuery };
var SetGraphLoading = (function () {
    function SetGraphLoading(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = SET_ACTIONS_GRAPH_LOADING;
    }
    return SetGraphLoading;
}());
export { SetGraphLoading };
var AddGraphQuery = (function () {
    function AddGraphQuery(payload) {
        this.payload = payload;
        this.type = ADD_ACTIONS_GRAPH_QUERY;
    }
    return AddGraphQuery;
}());
export { AddGraphQuery };
var SetSummary = (function () {
    function SetSummary(payload) {
        this.payload = payload;
        this.type = SET_ACTIONS_SUMMARY;
    }
    return SetSummary;
}());
export { SetSummary };
var AddSummaryQuery = (function () {
    function AddSummaryQuery(payload) {
        this.payload = payload;
        this.type = ADD_ACTIONS_SUMMARY_QUERY;
    }
    return AddSummaryQuery;
}());
export { AddSummaryQuery };
var SetSummaryQuery = (function () {
    function SetSummaryQuery(payload) {
        this.payload = payload;
        this.type = SET_ACTIONS_SUMMARY_QUERY;
    }
    return SetSummaryQuery;
}());
export { SetSummaryQuery };
var SetSummaryLoading = (function () {
    function SetSummaryLoading(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = SET_ACTIONS_SUMMARY_LOADING;
    }
    return SetSummaryLoading;
}());
export { SetSummaryLoading };
var SetSummaryActive = (function () {
    function SetSummaryActive(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = SET_ACTIONS_SUMMARY_ACTIVE;
    }
    return SetSummaryActive;
}());
export { SetSummaryActive };
