// ht-client/dist/es/dispatchers/loading-dispatcher.js
export var SET_USER_ANALYTICS = "[LOADING] set user analytics";
export var SET_USER_INDEX = "[LOADING] set user index";
export var SET_USER_ANALYTICS_ALL = "[LOADING] set user analytics all";
export var SET_USER_INDEX_ALL = "[LOADING] set user index all";
var SetLoadingUserAnalytics = (function () {
    function SetLoadingUserAnalytics(payload) {
        this.payload = payload;
        this.type = SET_USER_ANALYTICS;
    }
    return SetLoadingUserAnalytics;
}());
export { SetLoadingUserAnalytics };
var SetLoadingUserIndex = (function () {
    function SetLoadingUserIndex(payload) {
        this.payload = payload;
        this.type = SET_USER_INDEX;
    }
    return SetLoadingUserIndex;
}());
export { SetLoadingUserIndex };
var SetLoadingUserAnalyticsAll = (function () {
    function SetLoadingUserAnalyticsAll(payload) {
        this.payload = payload;
        this.type = SET_USER_ANALYTICS_ALL;
    }
    return SetLoadingUserAnalyticsAll;
}());
export { SetLoadingUserAnalyticsAll };
var SetLoadingUserIndexAll = (function () {
    function SetLoadingUserIndexAll(payload) {
        this.payload = payload;
        this.type = SET_USER_INDEX_ALL;
    }
    return SetLoadingUserIndexAll;
}());
export { SetLoadingUserIndexAll };
