// ht-client/dist/es/global/client.js
import { htRequestService } from "ht-api";
var HtClient = (function () {
    function HtClient(token, request) {
        if (token === void 0) { token = ""; }
        htRequestService.setInstance(request);
        this.token = token;
    }
    Object.defineProperty(HtClient.prototype, "token", {
        set: function (token) {
            htRequestService.getInstance().tokenServie.token = token;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HtClient.prototype, "tempToken", {
        set: function (token) {
            htRequestService.getInstance().tokenServie.tempToken = token;
        },
        enumerable: true,
        configurable: true
    });
    return HtClient;
}());
export { HtClient };
export var initClient = function (token, request) {
    return htClientService.getInstance(token, request);
};
export var htClientFactory = function (token, request) {
    return new HtClient(token, request);
};
export var htClientService = (function () {
    var instance;
    return {
        getInstance: function (token, request) {
            if (token === void 0) { token = ""; }
            if (!instance) {
                instance = htClientFactory(token, request);
            }
            return instance;
        }
    };
})();
