// ht-client/dist/es/global/date-range.js
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
import { BehaviorSubject } from "rxjs/BehaviorSubject";
import { map } from "rxjs/operators";
import { DateRangeMap } from "ht-data";
import { dateRangeDisplay } from "ht-utility";
export var defaultDateRange = __assign({}, DateRangeMap.today);
var DateRange = (function () {
    function DateRange(initialDate) {
        if (initialDate === void 0) { initialDate = {}; }
        this.dataBehaviour$ = new BehaviorSubject(__assign({}, defaultDateRange, initialDate));
        this.data$ = this.dataBehaviour$.asObservable();
    }
    ;
    DateRange.prototype.setDateRange = function (dateRange) {
        if (dateRange === void 0) { dateRange = null; }
        this.dataBehaviour$.next(dateRange);
    };
    DateRange.prototype.getDateRange = function () {
        return this.dataBehaviour$.getValue();
    };
    Object.defineProperty(DateRange.prototype, "display$", {
        get: function () {
            return this.data$.pipe(map(function (range) {
                return dateRangeDisplay(range);
            }));
        },
        enumerable: true,
        configurable: true
    });
    return DateRange;
}());
export { DateRange };
export var dateRangeFactory = function (initialDate) {
    if (initialDate === void 0) { initialDate = {}; }
    return new DateRange(initialDate);
};
export var dateRangeService = (function () {
    var instance;
    return {
        getInstance: function (initialDate) {
            if (!instance) {
                instance = dateRangeFactory(initialDate);
            }
            return instance;
        }
    };
})();
