// ht-data/dist/lib/rx-helpers/data-intercepter.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var combineLatest_1 = require("rxjs/observable/combineLatest");
var operators_1 = require("rxjs/operators");
var item_as_page_1 = require("./item-as-page");
exports.dataWithSelectedId$ = function (data$, id$, keys, prop) {
    return combineLatest_1.combineLatest(data$, id$, function (data, id) {
        if (!data && keys.length)
            return data;
        var allItems = keys.reduce(function (items, key) {
            return items.concat(data[key]);
        }, []);
        var selected = allItems.find(function (item) {
            return item.id === id;
        });
        return __assign({}, data, (_a = {}, _a[prop] = selected, _a));
        var _a;
    });
};
exports.listwithSelectedId$ = function (list$, id$) {
    return combineLatest_1.combineLatest(list$, id$.pipe(operators_1.distinctUntilChanged()), function (list, id) {
        if (!list)
            return list;
        return !id
            ? list
            : {
                count: 1,
                next: null,
                page: null,
                results: list.results.filter(function (item) {
                    return item.id === id;
                })
            };
    });
};
exports.listWithItem$ = function (list$, item$) {
    return combineLatest_1.combineLatest(list$, item$.pipe(item_as_page_1.itemAsPage$()), function (list, itemPage) {
        return itemPage ? itemPage : list;
    });
};
