// ht-client/dist/lib/global/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ht_api_1 = require("ht-api");
var HtClient = (function () {
    function HtClient(token, request) {
        if (token === void 0) { token = ""; }
        ht_api_1.htRequestService.setInstance(request);
        this.token = token;
    }
    Object.defineProperty(HtClient.prototype, "token", {
        set: function (token) {
            ht_api_1.htRequestService.getInstance().tokenServie.token = token;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HtClient.prototype, "tempToken", {
        set: function (token) {
            ht_api_1.htRequestService.getInstance().tokenServie.tempToken = token;
        },
        enumerable: true,
        configurable: true
    });
    return HtClient;
}());
exports.HtClient = HtClient;
exports.initClient = function (token, request) {
    return exports.htClientService.getInstance(token, request);
};
exports.htClientFactory = function (token, request) {
    return new HtClient(token, request);
};
exports.htClientService = (function () {
    var instance;
    return {
        getInstance: function (token, request) {
            if (token === void 0) { token = ""; }
            if (!instance) {
                instance = exports.htClientFactory(token, request);
            }
            return instance;
        }
    };
})();
