// ht-client/dist/lib/global/date-range.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var BehaviorSubject_1 = require("rxjs/BehaviorSubject");
var operators_1 = require("rxjs/operators");
var ht_data_1 = require("ht-data");
var ht_utility_1 = require("ht-utility");
exports.defaultDateRange = __assign({}, ht_data_1.DateRangeMap.today);
var DateRange = (function () {
    function DateRange(initialDate) {
        if (initialDate === void 0) { initialDate = {}; }
        this.dataBehaviour$ = new BehaviorSubject_1.BehaviorSubject(__assign({}, exports.defaultDateRange, initialDate));
        this.data$ = this.dataBehaviour$.asObservable();
    }
    ;
    DateRange.prototype.setDateRange = function (dateRange) {
        if (dateRange === void 0) { dateRange = null; }
        this.dataBehaviour$.next(dateRange);
    };
    DateRange.prototype.getDateRange = function () {
        return this.dataBehaviour$.getValue();
    };
    Object.defineProperty(DateRange.prototype, "display$", {
        get: function () {
            return this.data$.pipe(operators_1.map(function (range) {
                return ht_utility_1.dateRangeDisplay(range);
            }));
        },
        enumerable: true,
        configurable: true
    });
    return DateRange;
}());
exports.DateRange = DateRange;
exports.dateRangeFactory = function (initialDate) {
    if (initialDate === void 0) { initialDate = {}; }
    return new DateRange(initialDate);
};
exports.dateRangeService = (function () {
    var instance;
    return {
        getInstance: function (initialDate) {
            if (!instance) {
                instance = exports.dateRangeFactory(initialDate);
            }
            return instance;
        }
    };
})();
