// ht-maps/dist/es/entities/actions-heatmap.js
import { HeatmapMixin } from "../mixins/heatmap";
import { DataObservableMixin } from "../mixins/data-observable";
import { StyleMixin } from "../mixins/styles";
import { ExtendBoundsMixin } from "../mixins/extend-bounds";
import { MapItemsMixin } from "../mixins/map-items";
var ActionsHeatmap = (function () {
    function ActionsHeatmap(mapInstance) {
        this.mapInstance = mapInstance;
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case "google": {
                        return {
                            default: {}
                        };
                    }
                    case "leaflet": {
                        return {
                            default: {
                                radius: 7,
                                minOpacity: 0.6
                            }
                        };
                    }
                }
            }
        };
    }
    ActionsHeatmap.prototype.getPosition = function (item) {
        return {
            lat: item.completed_place__location[1],
            lng: item.completed_place__location[0],
        };
    };
    ;
    return ActionsHeatmap;
}());
export { ActionsHeatmap };
;
export var ActionsHeatmapTrace = ExtendBoundsMixin(DataObservableMixin(ExtendBoundsMixin(HeatmapMixin(StyleMixin(MapItemsMixin(ActionsHeatmap))))));
