// ht-maps/dist/es/entities/actions-polyline.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { AnimationMixin } from "../mixins/animation-renderer";
import { MarkersMixin } from "../mixins/marker-renderer";
import { PolylinesMixin } from "../mixins/polyline-renderer";
import { ExtendBoundsMixin } from "../mixins/extend-bounds";
import { SegmentPolylines } from "./segment-polylines";
import { TraceMixin } from "../mixins/trace";
import { SingleItemMixin } from "../mixins/single-item";
import { StyleMixin } from "../mixins/styles";
import { htAction } from "ht-data";
import { Color } from "ht-utility";
import { MapItemsMixin } from "../mixins/map-items";
var ActionsPolyline = (function (_super) {
    __extends(ActionsPolyline, _super);
    function ActionsPolyline() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.lineSymbol = {
            path: 'M 0,-1 0,1',
            strokeOpacity: 1,
            scale: 4
        };
        _this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case "google":
                        {
                            return {
                                default: {
                                    strokeColor: Color.grey3,
                                    strokeOpacity: 0,
                                    strokeWeight: 5,
                                    icons: [{
                                            icon: {
                                                path: 'M 0,-1 0,0',
                                                strokeOpacity: 1,
                                                scale: 4
                                            },
                                            offset: '0',
                                            repeat: '13px'
                                        }],
                                },
                                highlight: {
                                    strokeColor: Color.grey2,
                                    strokeOpacity: 1,
                                    strokeWeight: 5
                                },
                                fade: {
                                    strokeColor: Color.grey2,
                                    strokeOpacity: 0.2,
                                    strokeWeight: 2
                                }
                            };
                        }
                        ;
                    case "leaflet": {
                        return {
                            default: {
                                weight: 5,
                                color: Color.grey4,
                                opacity: 1,
                                dashArray: "7 10"
                            },
                            highlight: {
                                weight: 5,
                                color: Color.grey2,
                                opacity: 1
                            },
                            fade: {
                                weight: 2,
                                color: Color.grey2,
                                opacity: 0.2
                            }
                        };
                    }
                }
            },
        };
        _this.name = "action polyline";
        return _this;
    }
    ActionsPolyline.prototype.getPathFromData = function (data) {
        return data ? data.actions.reduce(function (path, action) {
            return htAction(action).getExpectedPosition() && !action.completed_at ? path.concat([htAction(action).getPosition()]) : path;
        }, []) : null;
    };
    return ActionsPolyline;
}(SegmentPolylines));
export { ActionsPolyline };
export var ActionsPolylineTrace = ConnectorMixin(AnimationMixin(SingleItemMixin(TraceMixin(ExtendBoundsMixin(PolylinesMixin(MarkersMixin(StyleMixin(MapItemsMixin(ActionsPolyline)))))))));
export function ConnectorMixin(Base) {
    return (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.positionTimeArray = [];
            return _this;
        }
        /*
        Need to ensure connector is traced before this class is traced
         */
        class_1.prototype.setConnector = function (entity) {
            this.connector = entity ? entity.item : null;
        };
        class_1.prototype.getConnectorPosition = function () {
            return this.connector ? this.mapInstance.mapUtils.getItemPosition(this.connector) : null;
        };
        class_1.prototype.update = function (_a, pathBearing) {
            var item = _a.item, data = _a.data;
            var startPosition;
            if (pathBearing) {
                var path = pathBearing.path;
                startPosition = path[path.length - 1];
            }
            else {
                startPosition = this.getConnectorPosition();
            }
            ;
            var dataPath = this.getPathFromData(data);
            if (dataPath && startPosition) {
                var finalPath = [startPosition].concat(dataPath);
                this.mapInstance.mapUtils.setPath(item, finalPath);
                this.mapInstance.mapUtils.setMap(item, this.mapInstance.map);
            }
            else {
                this.clear();
            }
        };
        return class_1;
    }(Base));
}
