// ht-maps/dist/es/entities/action-user.js
import { htUser } from "ht-data";
import { DivMarkersMixin } from "../mixins/div-markers-renderes";
import { TraceMixin } from "../mixins/trace";
import { MarkersMixin } from "../mixins/marker-renderer";
import { StyleMixin } from "../mixins/styles";
import { DataObservableMixin } from "../mixins/data-observable";
import { ExtendBoundsMixin } from "../mixins/extend-bounds";
import { AnimationsEntitiesMixin } from "../mixins/animations-entities";
import { MapItemsMixin } from "../mixins/map-items";
var ActionUser = (function () {
    function ActionUser(mapInstance) {
        this.mapInstance = mapInstance;
        this.name = "action user";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case "google":
                        {
                            return {
                                default: {
                                    flat: true,
                                    anchor: RichMarkerPosition.BOTTOM_CENTER
                                }
                            };
                        }
                        ;
                    case "leaflet": {
                        return {
                            default: {
                                zIndexOffset: 10,
                                iconSize: [60, 60],
                                className: "action-marker"
                                // iconAnchor: point(15, 43)
                                // iconAnchor: [0, 0]
                            }
                        };
                    }
                }
            }
        };
    }
    ActionUser.prototype.getPosition = function (data) {
        return htUser(data.user).getPosition();
    };
    ;
    ActionUser.prototype.trackBy = function (action) {
        return action.user.id;
    };
    ActionUser.prototype.trackAnimationBy = function (action) {
        return action.id;
    };
    ActionUser.prototype.getDivContent = function (data) {
        var content = "\n    <div style=\"border-radius: 50%; height: 60px; width: 60px; background: rgba(95,143,213,0.67) \">\n  <div style=\"height: 30px; width: 30px; background-image: url('" + data.user.photo + "'); background-repeat: no-repeat;\n  background-size: cover;     top: 15px;\n    position: relative;\n    left: 15px;\n    border-radius: 50%;\"></div>\n</div> \n    ";
        return content;
    };
    ;
    return ActionUser;
}());
export { ActionUser };
export var ActionUserTrace = DataObservableMixin(ExtendBoundsMixin(AnimationsEntitiesMixin(DivMarkersMixin(TraceMixin(MarkersMixin(StyleMixin(MapItemsMixin(ActionUser))))))));
