// ht-maps/dist/es/entities/action-markers.js
import { htAction } from "ht-data";
import { MarkersMixin } from "../mixins/marker-renderer";
import { StyleMixin } from "../mixins/styles";
import { TraceMixin } from "../mixins/trace";
import { ExtendBoundsMixin } from "../mixins/extend-bounds";
import { DivMarkersMixin } from "../mixins/div-markers-renderes";
import { NameCase } from "ht-utility";
import { MapItemsMixin } from "../mixins/map-items";
var ActionMarkers = (function () {
    function ActionMarkers(mapInstance) {
        this.mapInstance = mapInstance;
        this.name = "Action";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case 'google': {
                        return {
                            default: {
                                flat: true,
                                anchor: RichMarkerPosition.MIDDLE,
                                zIndex: 1
                            },
                            popup: {
                                pixelOffset: new google.maps.Size(0, -5),
                            }
                        };
                    }
                    case 'leaflet': {
                        return {
                            default: {
                                iconAnchor: [12, 12]
                                // iconSize: [35, 35],
                                // className: 'current-action-marker',
                                // iconAnchor: point(15, 43)
                                // iconAnchor: [15, 43]
                            },
                            popup: {
                                // offset: point(0, -35),
                                offset: [0, -5],
                                closeButton: false
                            }
                        };
                    }
                }
            }
        };
    }
    ActionMarkers.prototype.getPosition = function (data) {
        var posObj = htAction(data).getPositionsObject();
        return posObj ? posObj.position : null;
    };
    ActionMarkers.prototype.getDivContent = function (action) {
        var icon = "<div class=\"action-marker flex-row\">\n<span style=\"margin: auto\">" + NameCase(action.type[0]) + "</span>\n</div>";
        return icon;
    };
    return ActionMarkers;
}());
export { ActionMarkers };
export var ActionMarkersTrace = ExtendBoundsMixin(TraceMixin(DivMarkersMixin(MarkersMixin(StyleMixin(MapItemsMixin(ActionMarkers))))));
