// ht-maps/dist/es/entities/current-user.js
import { htUser } from "ht-data";
import { userDivFactory } from "../helpers/user-div-factory";
import { itemsFactory } from "../base/map-items-factory";
import { SingleItemMixin } from "../mixins/single-item";
import { DivMarkersMixin } from "../mixins/div-markers-renderes";
import { TraceMixin } from "../mixins/trace";
import { MarkersMixin } from "../mixins/marker-renderer";
import { StyleMixin } from "../mixins/styles";
import { AnimationMixin } from "../mixins/animation-renderer";
import { MapItemsMixin } from "../mixins/map-items";
export var currentUserConfig = {
    renderConfig: {
        getPosition: function (data) {
            return htUser(data).getPosition();
        },
        getDivContent: function (data) {
            return userDivFactory(data);
        }
    },
    styleFunct: {
        get: function (type) {
            switch (type) {
                case "google":
                    {
                        return {
                            default: {
                                flat: true,
                                anchor: RichMarkerPosition.BOTTOM_CENTER
                            }
                        };
                    }
                    ;
                case "leaflet": {
                    return {
                        default: {
                            // iconAnchor: point(15, 43)
                            iconAnchor: [15, 43]
                        }
                    };
                }
            }
        }
    },
    typeConfig: {
        isSingleItem: true,
        isDiv: true,
        hasDataObservable: false
    },
    name: "Current user"
};
export var currentUserTrace = function () {
    return itemsFactory(currentUserConfig);
};
var CurrentUser = (function () {
    function CurrentUser(mapInstance) {
        this.mapInstance = mapInstance;
        this.name = "Current user";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case "google":
                        {
                            return {
                                default: {
                                    flat: true,
                                    anchor: RichMarkerPosition.BOTTOM_CENTER
                                }
                            };
                        }
                        ;
                    case "leaflet": {
                        return {
                            default: {
                                // iconAnchor: point(15, 43)
                                iconAnchor: [15, 43]
                            }
                        };
                    }
                }
            }
        };
    }
    CurrentUser.prototype.getPosition = function (data) {
        return htUser(data).getPosition();
    };
    ;
    CurrentUser.prototype.getDivContent = function (data) {
        return userDivFactory(data);
    };
    return CurrentUser;
}());
export { CurrentUser };
export var CurrentUserTrace = AnimationMixin(SingleItemMixin(DivMarkersMixin(TraceMixin(MarkersMixin(StyleMixin(MapItemsMixin(CurrentUser)))))));
// export class CurrentUser {
//   name = "Current user";
//   styleFunct: StyleFunct = {
//     google: {
//       default: {
//         flat: true,
//         anchor: RichMarkerPosition.BOTTOM_CENTER,
//       }
//     },
//     leaflet: {
//       default: {
//
//       }
//     }
//   };
//
//   getPosition(data): HtPosition {
//     return htUser(data).getPosition()
//   };
//
//   getDivContent(data) {
//     return userDivFactory(data)
//
//   }
// };
//
// export const CurrentUserTrace = mapItemsFactory(CurrentUser, {
//   isSingleItem: true,
//   isDiv: true,
//   hasDataObservable: false
// });
