// ht-maps/dist/es/entities/destination-marker.js
import { ExtendBoundsMixin } from "../mixins/extend-bounds";
import { TraceMixin } from "../mixins/trace";
import { MarkersMixin } from "../mixins/marker-renderer";
import { StyleMixin } from "../mixins/styles";
import { DataObservableMixin } from "../mixins/data-observable";
import { DivMarkersMixin } from "../mixins/div-markers-renderes";
import { MapItemsMixin } from "../mixins/map-items";
var DestinationMap = (function () {
    function DestinationMap(mapInstance) {
        this.name = "Destination";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case 'google': {
                        return {
                            default: {},
                            popup: {
                                pixelOffset: new google.maps.Size(0, -5),
                            }
                        };
                    }
                    case 'leaflet': {
                        return {
                            default: {
                                zIndexOffset: 10,
                                iconSize: [30, 30],
                                className: "destination-marker"
                            }
                        };
                    }
                }
            }
        };
        this.mapInstance = mapInstance;
    }
    DestinationMap.prototype.getPosition = function (data) {
        var place = data.completed_place || data.expected_place;
        if (place && place.location) {
            return { lat: place.location.coordinates[1], lng: place.location.coordinates[0] };
        }
        else {
            return null;
        }
    };
    DestinationMap.prototype.getDivContent = function () {
        return "<div style=\"width: 30px; height: 30px; background: #00C94B; border-radius: 50%\"></div>";
    };
    return DestinationMap;
}());
export { DestinationMap };
;
export var DestinationMarker = DataObservableMixin(ExtendBoundsMixin(TraceMixin(DivMarkersMixin(MarkersMixin(StyleMixin(MapItemsMixin(DestinationMap)))))));
