// ht-maps/dist/es/entities/segment-polylines.js
import { segmentPolylineStyles } from "../styles/segment-polyline-styles";
import { ExtendBoundsMixin } from "../mixins/extend-bounds";
import { MarkersMixin } from "../mixins/marker-renderer";
import { StyleMixin } from "../mixins/styles";
import { TraceMixin } from "../mixins/trace";
import { PolylinesMixin } from "../mixins/polyline-renderer";
import { MapItemsMixin } from "../mixins/map-items";
var SegmentPolylines = (function () {
    function SegmentPolylines(mapInstance) {
        this.mapInstance = mapInstance;
        this.styleFunct = segmentPolylineStyles;
        this.name = "segment polyline";
    }
    SegmentPolylines.prototype.getEncodedPath = function (data) {
        return data.encoded_polyline;
    };
    ;
    SegmentPolylines.prototype.getEncodedPositionTime = function (data) {
        return data.time_aware_polyline;
    };
    SegmentPolylines.prototype.getPosition = function () {
        return { lat: 0, lng: 0 };
    };
    return SegmentPolylines;
}());
export { SegmentPolylines };
export var SegmentPolylinesTrace = TraceMixin(ExtendBoundsMixin(PolylinesMixin(MarkersMixin(StyleMixin(MapItemsMixin(SegmentPolylines))))));
// export class SegmentPolylines {
//   styleFunct = segmentPolylineStyles;
//
//   getEncodedPath(data) {
//     return data.encoded_polyline;
//   };
//
//   getEncodedPositionTime(data: ISegment) {
//     return data.time_aware_polyline
//   }
// }
//
// export const SegmentPolylinesTrace = mapItemsFactory(SegmentPolylines, {
//   isPolyline: true,
//   hasDataObservable: false
// });
// export const SegmentPolylinesTrace = _.compose(
//   PolylinesMixin,
//   MarkersMixin,
//   StyleMixin,
//   TraceMixin,
// )(SegmentPolylines);
