// ht-maps/dist/es/entities/start-marker.js
import { ExtendBoundsMixin } from "../mixins/extend-bounds";
import { TraceMixin } from "../mixins/trace";
import { MarkersMixin } from "../mixins/marker-renderer";
import { StyleMixin } from "../mixins/styles";
import { DataObservableMixin } from "../mixins/data-observable";
import { DivMarkersMixin } from "../mixins/div-markers-renderes";
import { MapItemsMixin } from "../mixins/map-items";
var StartMarker = (function () {
    function StartMarker(mapInstance) {
        this.name = "Start marker";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case 'google': {
                        return {
                            default: {},
                            popup: {
                                pixelOffset: new google.maps.Size(0, -5),
                            }
                        };
                    }
                    case 'leaflet': {
                        return {
                            default: {
                                zIndexOffset: 10,
                                iconSize: [20, 20],
                                className: "destination-marker"
                            }
                        };
                    }
                }
            }
        };
        this.mapInstance = mapInstance;
    }
    StartMarker.prototype.trace = function (mapItem) {
    };
    StartMarker.prototype.getPosition = function (data) {
        var place = data.started_place;
        if (place && place.location) {
            return { lat: place.location.coordinates[1], lng: place.location.coordinates[0] };
        }
        else {
            return null;
        }
    };
    StartMarker.prototype.getDivContent = function () {
        return "<div style=\"width: 20px; height: 20px; background: #9013FE; border-radius: 50%\"></div>";
    };
    return StartMarker;
}());
export { StartMarker };
;
export var StartMarkerTrace = ExtendBoundsMixin(DataObservableMixin(TraceMixin(DivMarkersMixin(MarkersMixin(StyleMixin(MapItemsMixin(StartMarker)))))));
