// ht-maps/dist/es/entities/stops-heatmap.js
import { HeatmapMixin } from "../mixins/heatmap";
import { DataObservableMixin } from "../mixins/data-observable";
import { StyleMixin } from "../mixins/styles";
import { ExtendBoundsMixin } from "../mixins/extend-bounds";
import { MapItemsMixin } from "../mixins/map-items";
var StopsHeatmap = (function () {
    function StopsHeatmap(mapInstance) {
        this.name = "stop heatmap";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case "google": {
                        return {
                            default: {}
                        };
                    }
                    case "leaflet": {
                        return {
                            default: {
                                radius: 7,
                                minOpacity: 0.6
                            }
                        };
                    }
                }
            }
        };
        this.mapInstance = mapInstance;
    }
    StopsHeatmap.prototype.getPosition = function (item) {
        return {
            lat: item.place__location[1],
            lng: item.place__location[0],
            weight: item.intensity
        };
    };
    return StopsHeatmap;
}());
export { StopsHeatmap };
;
export var StopsHeatmapTrace = ExtendBoundsMixin(DataObservableMixin(HeatmapMixin(StyleMixin(MapItemsMixin(StopsHeatmap)))));
