// ht-maps/dist/es/entities/users-cluster.js
import { htUser } from "ht-data";
import { userDivFactory } from "../helpers/user-div-factory";
import { ExtendBoundsMixin } from "../mixins/extend-bounds";
import { MarkersMixin } from "../mixins/marker-renderer";
import { StyleMixin } from "../mixins/styles";
import { DataObservableMixin } from "../mixins/data-observable";
import { PopupMixin } from "../mixins/popup-renderer";
import { DivMarkersMixin } from "../mixins/div-markers-renderes";
import { TraceMixin } from "../mixins/trace";
import { ClusterMixin } from "../mixins/clusters";
import { MapItemsMixin } from "../mixins/map-items";
var UsersCluster = (function () {
    function UsersCluster(mapInstance) {
        this.mapInstance = mapInstance;
        this.name = "users cluster";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case 'google': {
                        return {
                            default: {
                                flat: true,
                                anchor: RichMarkerPosition.BOTTOM_CENTER,
                                zIndex: 1
                            },
                            popup: {
                                disableAutoPan: true,
                                pixelOffset: new google.maps.Size(0, -35)
                            }
                        };
                    }
                    case 'leaflet': {
                        return {
                            default: {
                                // iconAnchor: point(15, 43)
                                iconAnchor: [15, 43]
                            },
                            popup: {
                                // offset: point(0, -35),
                                offset: [0, -35],
                                closeButton: false
                            }
                        };
                    }
                }
            }
        };
    }
    UsersCluster.prototype.getPosition = function (data) {
        return htUser(data).getPosition();
    };
    ;
    UsersCluster.prototype.getDivContent = function (data) {
        return userDivFactory(data);
    };
    ;
    UsersCluster.prototype.getInfoContent = function (data) {
        var string = "<div>\n<strong>" + data.name + "</strong>\n<div>" + data.display.status_text + "</div>\n<div>" + data.display.sub_status_text + "</div>\n</div>";
        return string;
    };
    UsersCluster.prototype.traceEffect = function () {
    };
    return UsersCluster;
}());
export { UsersCluster };
export var UsersClusterTrace = DataObservableMixin(PopupMixin(ClusterMixin(TraceMixin(ExtendBoundsMixin(DivMarkersMixin(MarkersMixin(StyleMixin(MapItemsMixin(UsersCluster)))))))));
// export class UsersClusters {
//   name = "Cluster user";
//   styleFunct: StyleFunct = {
//     google: {
//       default: {
//         flat: true,
//         anchor: RichMarkerPosition.BOTTOM_CENTER,
//         zIndex: 1
//       },
//       popup: {
//         disableAutoPan: true,
//         pixelOffset: new google.maps.Size(0, -35)
//       }
//     },
//     leaflet: {
//       default: {
//
//       }
//     }
//   };
//
//   getPosition(data): HtPosition {
//     return htUser(data).getPosition()
//   };
//
//   getDivContent(data) {
//     return userDivFactory(data)
//   };
//
//   getInfoContent(data) {
//     let string = `<div>
// <strong>${data.name}</strong>
// <div>${data.display.status_text}</div>
// <div>${data.display.sub_status_text}</div>
// </div>`;
//     return string
//   }
// }
//
// export const UsersClustersTrace = mapItemsFactory(UsersClusters, {
//   hasPopup: true,
//   isCluster: true,
//   isDiv: true,
//   hasDataObservable: true
// });
