// ht-client/dist/es/store/index.js
import { combineReducers, createReducerFactory } from "./utils";
import { Dispatcher } from "./dispatcher";
import { ReducerManager } from "./reducer-manager";
import { ScannedActionsSubject } from "./scanned-action";
import { State } from "./state";
import { Store } from "./store";
import { Effects } from "./effects";
var StoreProvider = (function () {
    function StoreProvider(reducers, config) {
        if (config === void 0) { config = {}; }
        this.INITIAL_STATE = config.initialState;
        this.metaReducers = config.metaReducers;
        this._REDUCER_FACTORY = config.reducerFactory
            ? config.reducerFactory
            : combineReducers;
        this.REDUCER_FACTORY = createReducerFactory(this._REDUCER_FACTORY, this.metaReducers);
        this.dispatcher = new Dispatcher(); //actionSubject
        this.EFFECTS_PROVIDERS = new Effects(this.dispatcher);
        this.reducerManager = new ReducerManager(this.dispatcher, this.INITIAL_STATE, reducers, this.REDUCER_FACTORY);
        this.SCANNED_ACTIONS_SUBJECT_PROVIDERS = new ScannedActionsSubject();
        this.STATE_PROVIDERS = new State(this.dispatcher, this.reducerManager, this.SCANNED_ACTIONS_SUBJECT_PROVIDERS, this.INITIAL_STATE);
        this.STORE_PROVIDERS = new Store(this.STATE_PROVIDERS, this.dispatcher, this.reducerManager);
    }
    return StoreProvider;
}());
export { StoreProvider };
export var storeFactory = function (reducers, config) {
    if (config === void 0) { config = {}; }
    var INITIAL_STATE = config.initialState;
    var metaReducers = config.metaReducers;
    var _REDUCER_FACTORY = config.reducerFactory
        ? config.reducerFactory
        : combineReducers;
    var REDUCER_FACTORY = createReducerFactory(_REDUCER_FACTORY, metaReducers);
    var dispatcher = new Dispatcher(); //actionSubject
    var EFFECTS_PROVIDERS = new Effects(dispatcher);
    var reducerManager = new ReducerManager(dispatcher, INITIAL_STATE, reducers, REDUCER_FACTORY);
    var SCANNED_ACTIONS_SUBJECT_PROVIDERS = new ScannedActionsSubject();
    var STATE_PROVIDERS = new State(dispatcher, reducerManager, SCANNED_ACTIONS_SUBJECT_PROVIDERS, INITIAL_STATE);
    var STORE_PROVIDERS = new Store(STATE_PROVIDERS, dispatcher, reducerManager);
    return {
        INITIAL_STATE: INITIAL_STATE,
        metaReducers: metaReducers,
        _REDUCER_FACTORY: _REDUCER_FACTORY,
        dispatcher: dispatcher,
        SCANNED_ACTIONS_SUBJECT_PROVIDERS: SCANNED_ACTIONS_SUBJECT_PROVIDERS,
        REDUCER_FACTORY: REDUCER_FACTORY,
        EFFECTS_PROVIDERS: EFFECTS_PROVIDERS,
        reducerManager: reducerManager,
        STATE_PROVIDERS: STATE_PROVIDERS,
        STORE_PROVIDERS: STORE_PROVIDERS
    };
};
