// ht-client/dist/lib/reducers/account-reducer.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var AccountDispatcher = require("../dispatchers/accounts-dispatcher");
var selector_1 = require("../store/selector");
exports.initialState = {
    key: "false"
};
function reducer(state, action) {
    if (state === void 0) { state = exports.initialState; }
    switch (action.type) {
        case AccountDispatcher.SET_ACCOUNT_USER: {
            return __assign({}, state, { accountUser: action.payload });
        }
        case AccountDispatcher.SET_MEMBERHSHIPS_ALL: {
            // let results = state.membershipsAll ? [...state.membershipsAll.results, ...action.payload.results] : action.payload.results;
            // return {...state, membershipsAll: {...action.payload, results}}
            return __assign({}, state, { membershipsAll: action.payload });
        }
        case AccountDispatcher.SET_KEY: {
            return __assign({}, state, { key: action.payload });
        }
        case AccountDispatcher.SET_TEMP_KEY: {
            return __assign({}, state, { tempKey: action.payload });
        }
        case AccountDispatcher.SET_USER_ID: {
            return __assign({}, state, { userId: action.payload });
        }
        default: {
            return state;
        }
    }
}
exports.reducer = reducer;
exports.getAccountUser = function (state) { return state.accountUser; };
exports.getMembershipsAll = function (state) { return state.membershipsAll; };
exports.getKey = function (state) { return state.key; };
exports.getTempKey = function (state) { return state.tempKey; };
exports.getUserId = function (state) { return state.userId; };
exports.getCurrentKey = selector_1.createSelector(exports.getKey, exports.getTempKey, function (key, tempKey) {
    return tempKey || key;
});
