// ht-client/dist/lib/reducers/user-reducer.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var UserDispatch = require("../dispatchers/user-dispatcher");
var interfaces_1 = require("../interfaces");
var selector_1 = require("../store/selector");
var _ = require("underscore");
var initialUsersAnalyticsAll = {
    resultsEntity: {},
    isFirst: false
};
var initialUsersIndexAll = {
    resultsEntity: {},
    isFirst: false
};
var initialState = {
    usersListActive: false,
    usersMarkersActive: false,
    usersSummaryActive: false,
    listApiType: interfaces_1.ApiType.analytics,
    usersAnalyticsAll: initialUsersAnalyticsAll,
    usersIndexAll: initialUsersIndexAll
};
function usersReducer(state, action) {
    if (state === void 0) { state = initialState; }
    switch (action.type) {
        /*
        placeline
         */
        case UserDispatch.SET_USER_DATA: {
            return __assign({}, state, { userData: action.payload });
        }
        case UserDispatch.SET_PLACELINE_ID: {
            return __assign({}, state, { placelineId: action.payload });
        }
        case UserDispatch.TOGGLE_PLACELINE_ID: {
            var placelineId = state.placelineId == action.payload ? null : action.payload;
            return __assign({}, state, { placelineId: placelineId });
        }
        case UserDispatch.SET_PLACELINE_QUERY: {
            return __assign({}, state, { placelineQuery: action.payload });
        }
        case UserDispatch.ADD_PLACELINE_QUERY: {
            return __assign({}, state, { placelineQuery: __assign({}, state.placelineQuery, action.payload) });
        }
        case UserDispatch.SET_PLACELINE_LOADING: {
            return __assign({}, state, { placelineLoading: !!action.payload });
        }
        /*
        List
         */
        case UserDispatch.SET_LIST_ID: {
            return __assign({}, state, { listId: action.payload });
        }
        case UserDispatch.TOGGLE_LIST_ID: {
            var listId = state.listId == action.payload ? null : action.payload;
            return __assign({}, state, { listId: listId });
        }
        case UserDispatch.SET_LIST_ACTIVE: {
            return __assign({}, state, { usersListActive: action.payload });
        }
        case UserDispatch.ADD_LIST_QUERY: {
            return __assign({}, state, { listQuery: __assign({}, state.listQuery, action.payload) });
        }
        case UserDispatch.SET_LIST_QUERY: {
            return __assign({}, state, { listQuery: action.payload });
        }
        case UserDispatch.CLEAR_QUERY_KEY: {
            var listQuery = __assign({}, state.listQuery);
            if (listQuery)
                delete listQuery[action.payload];
            return __assign({}, state, { listQuery: listQuery });
        }
        /*
        Analytics page
         */
        case UserDispatch.SET_USERS_ANALYTICS_PAGE: {
            return __assign({}, state, { usersAnalyticsPage: action.payload });
        }
        case UserDispatch.SET_USERS_ANALYTICS_LOADING: {
            return __assign({}, state, { analyticsLoading: action.payload });
        }
        case UserDispatch.SET_USERS_INDEX_PAGE: {
            return __assign({}, state, { usersIndexPage: action.payload });
        }
        case UserDispatch.ADD_USERS_ANALYTICS_ALL: {
            var resultsEntity = {};
            if (state.usersAnalyticsAll) {
                resultsEntity = __assign({}, state.usersAnalyticsAll.resultsEntity, _.indexBy(action.payload.results, "id"));
            }
            return __assign({}, state, { usersAnalyticsAll: {
                    next: action.payload.next,
                    previous: action.payload.previous,
                    count: action.payload.count,
                    resultsEntity: resultsEntity
                } });
        }
        case UserDispatch.SET_USERS_ANALYTICS_ALL: {
            var resultsEntity = {};
            if (state.usersAnalyticsAll) {
                resultsEntity = _.indexBy(action.payload.results, "id");
            }
            return __assign({}, state, { usersAnalyticsAll: {
                    next: action.payload.next,
                    previous: action.payload.previous,
                    count: action.payload.count,
                    resultsEntity: resultsEntity
                } });
        }
        case UserDispatch.SET_USERS_ANALYTICS_ALL_LOADING: {
            return __assign({}, state, { analyticsAllLoading: action.payload });
        }
        case UserDispatch.ADD_USERS_INDEX_ALL: {
            // const newEntities = _.indexBy(action.payload.results, 'id');
            var resultsEntity = state.usersIndexAll
                ? __assign({}, state.usersIndexAll.resultsEntity, action.payload.resultsEntity) : {};
            return __assign({}, state, { usersIndexAll: __assign({}, action.payload, { resultsEntity: resultsEntity }) });
        }
        case UserDispatch.SET_USERS_LIST_DATA_MAP: {
            return __assign({}, state, { usersListDataMap: action.payload });
        }
        case UserDispatch.SET_USERS_MARKERS_DATA_MAP: {
            return __assign({}, state, { usersMarkersDataMap: action.payload });
        }
        case UserDispatch.SET_USERS_LIST_API_TYPE: {
            return __assign({}, state, { listApiType: action.payload });
        }
        case UserDispatch.SET_MARKERS_ACTIVE: {
            return __assign({}, state, { usersMarkersActive: action.payload });
        }
        /*
        Summary
         */
        case UserDispatch.SET_SUMMARY_ACTIVE: {
            return __assign({}, state, { usersSummaryActive: action.payload });
        }
        case UserDispatch.SET_USERS_SUMMARY: {
            return __assign({}, state, { usersSummary: action.payload });
        }
        case UserDispatch.SET_USERS_SUMMARY_LOADING: {
            return __assign({}, state, { summaryLoading: !!action.payload });
        }
        default: {
            return state;
        }
    }
}
exports.usersReducer = usersReducer;
exports.getUserData = function (state) { return state.userData; };
exports.getPlacelineId = function (state) { return state.placelineId; };
exports.getPlacelineQuery = function (state) { return state.placelineQuery; };
exports.getPlacelineLoading = function (state) { return state.placelineLoading; };
exports.getListActive = function (state) { return state.usersListActive; };
exports.getListId = function (state) { return state.listId; };
exports.getListQuery = function (state) { return state.listQuery; };
exports.getAnalyticsPage = function (state) { return state.usersAnalyticsPage; };
exports.getAnalyticsLoading = function (state) { return state.analyticsLoading; };
exports.getAnalyticsAll = function (state) { return state.usersAnalyticsAll; };
exports.getAnalyticsAllLoading = function (state) {
    return state.analyticsAllLoading;
};
exports.getMarkerDataMap = function (state) { return state.usersMarkersDataMap; };
exports.getIndexAll = function (state) { return state.usersIndexAll; };
// export const getAnalyticFilteredsMarkers = (state: State) => validMarkers(state.usersAnalyticsAll);
// export const getIndexFilteredMarkers = (state: State) => validMarkers(state.usersIndexAll);
exports.getIndexPage = function (state) { return state.usersIndexPage; };
exports.getListApiType = function (state) { return state.listApiType; };
exports.getMarkersActive = function (state) { return state.usersMarkersActive; };
exports.getSummary = function (state) { return state.usersSummary; };
exports.getSummaryActive = function (state) { return state.usersSummaryActive; };
exports.getSummaryLoading = function (state) { return state.summaryLoading; };
exports.getAnalyticFilteredMarkers = selector_1.createSelector(exports.getAnalyticsAll, exports.getMarkerDataMap, function (allData, mapFunc) {
    return mapFunc ? mapFunc(allData) : allData;
});
exports.getIndexFilteredMarkers = selector_1.createSelector(exports.getIndexAll, exports.getMarkerDataMap, function (allData, mapFunc) {
    return mapFunc ? mapFunc(allData) : allData;
});
exports.getIndexActive = selector_1.createSelector(exports.getListApiType, exports.getListActive, function (apiType, isListActive) {
    return false;
    // return apiType === ApiType.index && isListActive
});
exports.getAnalyticsActive = selector_1.createSelector(exports.getListApiType, exports.getListActive, function (apiType, isListActive) {
    return isListActive;
    // return apiType === ApiType.analytics && isListActive
});
exports.getIndexMarkersActive = selector_1.createSelector(exports.getListApiType, exports.getMarkersActive, function (apiType, isMarkersActive) {
    return false;
    // return apiType === ApiType.index && isMarkersActive
});
exports.getAnalyticsMarkersActive = selector_1.createSelector(exports.getListApiType, exports.getMarkersActive, function (apiType, isMarkersActive) {
    return isMarkersActive;
    // return apiType === ApiType.analytics && isMarkersActive
});
