// ht-client/dist/es/base/item-client.js
var EntityItemClient = (function () {
    function EntityItemClient() {
        //listGetData
        this.updateStrategy = "live";
        this.pollDuration = 10000;
        // api$;
        this.name = "item";
        this.allowedQueryKeys = null;
        // setData(data) {
        //
        // };
        // setActive(isActive: boolean = true){
        //
        // }
    }
    // init: () => void;
    // getData$: (data) => any;
    EntityItemClient.prototype.firstDataEffect = function () {
        this.setLoading(false);
    };
    EntityItemClient.prototype.getDefaultQuery = function () {
        return {};
    };
    Object.defineProperty(EntityItemClient.prototype, "apiParams$", {
        get: function () {
            return this.getApiParams$();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(EntityItemClient.prototype, "apiQuery$", {
        get: function () {
            return this.getApiQuery$();
        },
        enumerable: true,
        configurable: true
    });
    return EntityItemClient;
}());
export { EntityItemClient };
