// ht-client/dist/lib/entities/actions/actions-graph.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var get_first_data_1 = require("../../mixins/get-first-data");
var entity_query_1 = require("../../mixins/entity-query");
var get_data_1 = require("../../mixins/get-data");
var client_subscription_1 = require("../../mixins/client-subscription");
var BehaviorSubject_1 = require("rxjs/BehaviorSubject");
var of_1 = require("rxjs/observable/of");
var ActionsGraph = (function () {
    function ActionsGraph(_a) {
        var dateRange = _a.dateRange, store = _a.store, dateParam = _a.dateParam, api = _a.api;
        this.query$ = of_1.of({});
        this.dataBehaviour$ = new BehaviorSubject_1.BehaviorSubject(null);
        this.data$ = this.dataBehaviour$.asObservable();
        this.loadingBehaviour$ = new BehaviorSubject_1.BehaviorSubject(false);
        this.loading$ = this.loadingBehaviour$.asObservable();
        this.updateStrategy = "once";
        this.pollDuration = 10000;
        this.api$ = function (query) { return api.graph(query); };
        this.dateRange = dateRange;
        this.dateParam = dateParam;
    }
    ActionsGraph.prototype.getDefaultQuery = function () {
        return {};
    };
    ;
    ActionsGraph.prototype.setData = function (data) {
        this.dataBehaviour$.next(data);
    };
    ActionsGraph.prototype.firstDataEffect = function () {
        this.setLoading(false);
    };
    ActionsGraph.prototype.setLoading = function (loading) {
        if (loading === void 0) { loading = true; }
        this.loadingBehaviour$.next(!!loading);
    };
    ActionsGraph.prototype.clearData = function () {
        this.setData(null);
    };
    ActionsGraph.prototype.destroy = function () {
        this.clearData();
        this.dataSub.unsubscribe();
    };
    return ActionsGraph;
}());
exports.ActionsGraph = ActionsGraph;
exports.ActionsGraphClient = client_subscription_1.clientSubMixin(get_data_1.getQueryDataMixin(get_first_data_1.getFirstDataMixin(entity_query_1.listQueryMixin(ActionsGraph))));
