// ht-client/dist/lib/entities/actions/actions-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var date_range_1 = require("../../global/date-range");
var actions_graph_1 = require("./actions-graph");
var fromActions = require("../../reducers/actions-reducer");
var store_provider_1 = require("../../global/store-provider");
var actions_list_client_1 = require("./actions-list-client");
var actions_summary_client_1 = require("./actions-summary-client");
var actions_filter_1 = require("../../filters/actions-filter");
var actions_heatmap_client_1 = require("./actions-heatmap-client");
var actions_list_all_client_1 = require("./actions-list-all-client");
var ht_api_1 = require("ht-api");
var HtActionsClient = (function () {
    function HtActionsClient(config) {
        this.filters = new actions_filter_1.ActionsFilter();
        var api = new ht_api_1.HtApi().actions;
        this.api = api;
        var store = store_provider_1.ApiStoreService.getNewInstance();
        store.addReducer("actions", fromActions.actionsReducer);
        this.store = store;
        var dateRange = config.dateRange;
        var dateParam = 'created_at';
        this.graph = new actions_graph_1.ActionsGraphClient({ dateRange: dateRange, dateParam: dateParam, api: api });
        this.list = new actions_list_client_1.ActionsListClient({ dateRange: dateRange, store: store, dateParam: dateParam, api: api });
        this.listAll = new actions_list_all_client_1.ActionsIndexAllClient({ dateRange: dateRange, dateParam: dateParam, api: api });
        this.summary = new actions_summary_client_1.ActionsSummaryClient({ dateRange: dateRange, store: store, dateParam: dateParam, api: api });
        this.heatmap = new actions_heatmap_client_1.ActionsHeatmapClient({ dateRange: dateRange, dateParam: 'completed_at', api: api });
    }
    return HtActionsClient;
}());
exports.HtActionsClient = HtActionsClient;
exports.actionsClientFactory = function (options) {
    if (options === void 0) { options = {}; }
    var dateRange = options.dateRange || date_range_1.dateRangeService.getInstance();
    return new HtActionsClient({ dateRange: dateRange });
};
