// ht-client/dist/lib/entities/actions/actions-heatmap-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var entity_query_1 = require("../../mixins/entity-query");
var list_all_client_sub_1 = require("../../mixins/list-all-client-sub");
var get_data_1 = require("../../mixins/get-data");
var get_first_data_1 = require("../../mixins/get-first-data");
var of_1 = require("rxjs/observable/of");
var BehaviorSubject_1 = require("rxjs/BehaviorSubject");
var ht_data_1 = require("ht-data");
var ActionsHeatmap = (function () {
    function ActionsHeatmap(_a) {
        var dateRange = _a.dateRange, dateParam = _a.dateParam, api = _a.api;
        this.query$ = of_1.of({});
        this.loadingState$ = new BehaviorSubject_1.BehaviorSubject(false);
        this.loading$ = this.loadingState$.asObservable();
        this.dataState$ = new BehaviorSubject_1.BehaviorSubject(null);
        this.data$ = this.dataState$.asObservable();
        this.active$ = new BehaviorSubject_1.BehaviorSubject(false);
        this.dataArray$ = this.data$.pipe(ht_data_1.PageResults$);
        this.api$ = function (query) { return api.allPages(api.heatmap(query)); };
        this.dateRange = dateRange;
        this.dateParam = dateParam;
    }
    ActionsHeatmap.prototype.setActive = function (active) {
        if (active === void 0) { active = true; }
        this.active$.next(active);
    };
    ActionsHeatmap.prototype.getDefaultQuery = function () {
        return { page_size: 300 };
    };
    ;
    ActionsHeatmap.prototype.firstDataEffect = function (data) {
        if ((data && !data.next) || !data) {
            this.setLoading(false);
        }
    };
    ActionsHeatmap.prototype.setLoading = function (isLoading) {
        this.loadingState$.next(isLoading);
    };
    ActionsHeatmap.prototype.setData = function (data) {
        this.dataState$.next(data);
    };
    ActionsHeatmap.prototype.addData = function (data) {
        this.dataState$.next(data);
    };
    return ActionsHeatmap;
}());
exports.ActionsHeatmap = ActionsHeatmap;
exports.ActionsHeatmapClient = list_all_client_sub_1.listAllClientSubMixin(get_data_1.getAllPageDataMixin(get_first_data_1.getFirstDataMixin(entity_query_1.listQueryMixin(ActionsHeatmap))));
