// ht-client/dist/lib/entities/actions/actions-summary-client.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var list_client_1 = require("../../base/list-client");
var fromRoot = require("../../reducers");
var fromActions = require("../../dispatchers/actions-dispatcher");
var entity_query_1 = require("../../mixins/entity-query");
var get_data_1 = require("../../mixins/get-data");
var get_first_data_1 = require("../../mixins/get-first-data");
var client_subscription_1 = require("../../mixins/client-subscription");
var actions_filter_1 = require("../../filters/actions-filter");
var operators_1 = require("rxjs/operators");
var ActionsSummary = (function (_super) {
    __extends(ActionsSummary, _super);
    function ActionsSummary(_a) {
        var dateRange = _a.dateRange, store = _a.store, dateParam = _a.dateParam, api = _a.api;
        var _this = _super.call(this) || this;
        _this.allowedQueryMap = [
            {
                key: "show_all",
            },
            {
                key: "search"
            }
        ];
        _this.updateStrategy = 'live';
        // updateStrategy = 'once';
        _this.filter = new actions_filter_1.ActionsFilter();
        _this.api$ = function (query) { return api.summary(query); };
        _this.dateRange = dateRange;
        _this.store = store;
        _this.dateParam = dateParam;
        _this.query$ = _this.store.select(fromRoot.getActionsSummaryQuery);
        _this.active$ = _this.store.select(fromRoot.getActionsSummaryActive);
        _this.data$ = _this.store.select(fromRoot.getActionsSummary);
        _this.summaryChart$ = _this.getSummaryChart();
        _this.loading$ = _this.store.select(fromRoot.getActionsSummaryLoading);
        return _this;
    }
    ActionsSummary.prototype.setActive = function (isActive) {
        if (isActive === void 0) { isActive = true; }
        isActive = isActive ? new Date().toISOString() : isActive;
        this.store.dispatch(new fromActions.SetSummaryActive(isActive));
    };
    ActionsSummary.prototype.setLoading = function (loading) {
        this.store.dispatch(new fromActions.SetSummaryLoading(loading));
    };
    ;
    ActionsSummary.prototype.setData = function (data) {
        this.store.dispatch(new fromActions.SetSummary(data));
    };
    ActionsSummary.prototype.setQuery = function (query) {
        this.store.dispatch(new fromActions.SetSummaryQuery(query));
    };
    ;
    ActionsSummary.prototype.getSummaryChart = function () {
        var _this = this;
        return this.data$.pipe(operators_1.map(function (summaryData) {
            return _this.filter.summaryCharts(_this.filter.statusQueryArray, summaryData);
        }));
    };
    return ActionsSummary;
}(list_client_1.EntityListClient));
exports.ActionsSummary = ActionsSummary;
;
exports.ActionsSummaryClient = client_subscription_1.clientSubMixin(get_data_1.getQueryDataMixin(get_first_data_1.getFirstDataMixin(entity_query_1.listQueryMixin(ActionsSummary))));
