// ht-client/dist/lib/entities/actions/actions-list-all-client.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var entity_query_1 = require("../../mixins/entity-query");
var get_first_data_1 = require("../../mixins/get-first-data");
var list_all_client_sub_1 = require("../../mixins/list-all-client-sub");
var get_data_1 = require("../../mixins/get-data");
var all_items_client_1 = require("../../base/all-items.client");
var BehaviorSubject_1 = require("rxjs/BehaviorSubject");
var ActionsIndexAll = (function (_super) {
    __extends(ActionsIndexAll, _super);
    function ActionsIndexAll(_a) {
        var dateRange = _a.dateRange, store = _a.store, dateParam = _a.dateParam, api = _a.api;
        var _this = _super.call(this) || this;
        _this.dataBehaviour$ = new BehaviorSubject_1.BehaviorSubject(null);
        _this.loadingBehaviour$ = new BehaviorSubject_1.BehaviorSubject(false);
        _this.loading$ = _this.loadingBehaviour$.asObservable();
        _this.activeBehaviour$ = new BehaviorSubject_1.BehaviorSubject(false);
        _this.api$ = function (query) { return api.allPages(api.index(query)); };
        _this.dateRange = dateRange;
        _this.dateParam = dateParam;
        _this.query$ = new BehaviorSubject_1.BehaviorSubject(_this.getDefaultQuery());
        _this.active$ = _this.activeBehaviour$.asObservable();
        return _this;
    }
    ActionsIndexAll.prototype.setActive = function (isActive) {
        if (isActive === void 0) { isActive = true; }
        this.activeBehaviour$.next(isActive);
    };
    ActionsIndexAll.prototype.setQuery = function (query) {
        this.query$.next(query);
    };
    Object.defineProperty(ActionsIndexAll.prototype, "data$", {
        get: function () {
            return this.dataBehaviour$.asObservable();
        },
        enumerable: true,
        configurable: true
    });
    ActionsIndexAll.prototype.addData = function (data) {
        this.dataBehaviour$.next(data);
    };
    ;
    ActionsIndexAll.prototype.setData = function (data) {
        this.dataBehaviour$.next(data);
    };
    ActionsIndexAll.prototype.setLoading = function (loading) {
    };
    return ActionsIndexAll;
}(all_items_client_1.EntityAllItemsClient));
exports.ActionsIndexAll = ActionsIndexAll;
;
exports.ActionsIndexAllClient = list_all_client_sub_1.listAllClientSubMixin(get_data_1.getAllPageDataMixin(get_first_data_1.getFirstDataMixin(entity_query_1.listQueryMixin(ActionsIndexAll))));
