// ht-client/dist/lib/entities/actions/actions-list-client.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var list_client_1 = require("../../base/list-client");
var fromRoot = require("../../reducers");
var ht_data_1 = require("ht-data");
var fromAction = require("../../dispatchers/actions-dispatcher");
var entity_query_1 = require("../../mixins/entity-query");
var get_first_data_1 = require("../../mixins/get-first-data");
var client_subscription_1 = require("../../mixins/client-subscription");
var get_data_1 = require("../../mixins/get-data");
var ActionsList = (function (_super) {
    __extends(ActionsList, _super);
    function ActionsList(_a) {
        var dateRange = _a.dateRange, store = _a.store, dateParam = _a.dateParam, api = _a.api;
        var _this = _super.call(this) || this;
        _this.api$ = function (query) {
            return api.index(query);
        };
        _this.dateRange = dateRange;
        _this.store = store;
        _this.query$ = _this.store.select(fromRoot.getActionsListQuery);
        _this.dateParam = dateParam;
        _this.active$ = _this.store.select(fromRoot.getActionsListActive);
        _this.data$ = _this.store.select(fromRoot.getActionsList);
        _this.loading$ = _this.store.select(fromRoot.getActionsListLoading);
        _this.dataArray$ = _this.data$.pipe(ht_data_1.PageResults$);
        return _this;
        // this.init()
    }
    ActionsList.prototype.setActive = function (active) {
        if (active === void 0) { active = true; }
        this.store.dispatch(new fromAction.SetListActive(active));
    };
    ActionsList.prototype.setLoading = function (loading) {
        this.store.dispatch(new fromAction.SetListLoading(loading));
    };
    ;
    ActionsList.prototype.setData = function (data) {
        this.store.dispatch(new fromAction.SetList(data));
    };
    ;
    ActionsList.prototype.setQuery = function (query) {
        this.store.dispatch(new fromAction.SetListQuery(query));
    };
    return ActionsList;
}(list_client_1.EntityListClient));
exports.ActionsList = ActionsList;
;
exports.ActionsListClient = client_subscription_1.clientSubMixin(get_data_1.getPageDataMixin(get_first_data_1.getFirstDataMixin(entity_query_1.listQueryMixin(ActionsList))));
