// ht-client/dist/lib/mixins/entity-query.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
// import {MergeQuery} from "ht-data";
var ht_data_1 = require("ht-data");
var operators_1 = require("rxjs/operators");
var combineLatest_1 = require("rxjs/observable/combineLatest");
var operators_2 = require("../helpers/operators");
var of_1 = require("rxjs/observable/of");
function listQueryMixin(Base) {
    return (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_1.prototype.getApiQuery$ = function () {
            return this.getApiParams$().pipe(operators_1.map(function (data) {
                return data[0];
            }));
        };
        class_1.prototype.getBaseQuery$ = function () {
            var baseQuery$ = this.query$.pipe(ht_data_1.AllowedQueryMap(this.allowedQueryMap), operators_2.MergeQuery(this.getDefaultQuery()), ht_data_1.CombineQueries([this.dateRange.data$.pipe(ht_data_1.DateRangeToQuery$(this.dateParam)) || of_1.of({})]));
            return baseQuery$;
        };
        class_1.prototype.getApiParams$ = function () {
            var baseQuery$ = this.getBaseQuery$().pipe(operators_1.map(function (data) {
                return [data];
            }));
            return this.active$
                ? this.active$.pipe(operators_1.switchMap(function (isActive) {
                    // console.log(isActive, "accc");
                    return isActive ? baseQuery$ : of_1.of([]);
                }))
                : baseQuery$;
        };
        return class_1;
    }(Base));
}
exports.listQueryMixin = listQueryMixin;
function itemQueryMixin(Base) {
    return (function (_super) {
        __extends(class_2, _super);
        function class_2() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_2.prototype.getApiQuery$ = function () {
            return this.getApiParams$().pipe(operators_1.map(function (data) {
                return data[1];
            }));
        };
        class_2.prototype.getApiParams$ = function () {
            return combineLatest_1.combineLatest(this.id$, this.query$.pipe(operators_2.MergeQuery(this.getDefaultQuery())));
        };
        return class_2;
    }(Base));
}
exports.itemQueryMixin = itemQueryMixin;
