// ht-client/dist/lib/base/list-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EntityListClient = (function () {
    function EntityListClient() {
        //listGetData
        this.updateStrategy = "live";
        this.pollDuration = 10000;
        // api$;
        this.name = "list";
        //
        // setData(data) {
        //
        // };
        //
        // setActive(isActive: boolean = true){
        //
        // }
    }
    // init: () => void;
    // getData$: (data) => any;
    // dataSub: Subscription;
    EntityListClient.prototype.firstDataEffect = function (data) {
        this.setLoading(false);
    };
    EntityListClient.prototype.getDefaultQuery = function () {
        return { page_size: 10 };
    };
    Object.defineProperty(EntityListClient.prototype, "apiParams$", {
        get: function () {
            return this.getApiParams$();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(EntityListClient.prototype, "apiQuery$", {
        get: function () {
            return this.getApiQuery$();
        },
        enumerable: true,
        configurable: true
    });
    return EntityListClient;
}());
exports.EntityListClient = EntityListClient;
