// ht-client/dist/lib/entities/account-user/accounts-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fromAccount = require("../../reducers/account-reducer");
var store_provider_1 = require("../../global/store-provider");
var account_user_client_1 = require("./account-user-client");
var memberships_all_client_1 = require("./memberships-all-client");
var fromAccounts = require("../../dispatchers/accounts-dispatcher");
var ht_api_1 = require("ht-api");
var AccountsClient = (function () {
    function AccountsClient() {
        var api = new ht_api_1.HtApi().accountUser;
        this.api = api;
        this.store = store_provider_1.ApiStoreService.getInstance();
        this.store.addReducer("accounts", fromAccount.reducer);
        this.accountUser = new account_user_client_1.AccountUserClient({ store: this.store, api: api });
        this.memberships = new memberships_all_client_1.MemberShipsClient({ store: this.store, api: api });
    }
    AccountsClient.prototype.setToken = function (token) {
        this.store.dispatch(new fromAccounts.SetKey(token));
    };
    return AccountsClient;
}());
exports.AccountsClient = AccountsClient;
