// time-aware-polyline/dist/lib/polyline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var time_aware_1 = require("./time-aware");
var PolylineUtils = /** @class */ (function () {
    function PolylineUtils() {
        this.encoder = new time_aware_1.TimeAwareEncoder();
    }
    PolylineUtils.prototype.updateTimeAwarePolyline = function (encodedPolyline) {
        if (this.isNewPolyline(encodedPolyline)) {
            this.encodedPolyline = encodedPolyline;
            this.timeAwarePolyline = this.encoder.decodeTimeAwarePolyline(this.encodedPolyline);
        }
    };
    PolylineUtils.prototype.getPolylineToTime = function (timestamp) {
        return this.encoder.getLocationsElapsedByTimestamp(this.timeAwarePolyline, timestamp);
    };
    PolylineUtils.prototype.getLatestTime = function () {
        if (this.timeAwarePolyline) {
            var lastI = this.timeAwarePolyline.length - 1;
            return lastI > -1 ? this.timeAwarePolyline[lastI][2] : null;
        }
        else {
            return null;
        }
    };
    PolylineUtils.prototype.isNewPolyline = function (encodedPolyline) {
        return encodedPolyline != this.encodedPolyline;
    };
    return PolylineUtils;
}());
exports.PolylineUtils = PolylineUtils;
