// ht-utility/dist/lib/date-range.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var date_fns_1 = require("date-fns");
var date_string_1 = require("./date-string");
exports.DateMapService = (function () {
    var instance;
    return {
        getInstance: function () {
            if (!instance) {
                instance = {
                    today_end: date_fns_1.endOfToday().toISOString(),
                    today_start: date_fns_1.startOfToday().toISOString(),
                    yesterday_start: date_fns_1.startOfYesterday().toISOString(),
                    yesterday_end: date_fns_1.endOfYesterday().toISOString(),
                    day_7_start: date_fns_1.startOfDay(date_fns_1.addDays(new Date(), -6)).toISOString(),
                    month_start: date_fns_1.startOfMonth(new Date()).toISOString(),
                    day_30_start: date_fns_1.startOfDay(date_fns_1.addDays(new Date(), -29)).toISOString(),
                };
            }
            return instance;
        }
    };
})();
exports.DateRangeMap = {
    "today": {
        start: exports.DateMapService.getInstance().today_start,
        end: exports.DateMapService.getInstance().today_end
    },
    "yesterday": {
        start: exports.DateMapService.getInstance().yesterday_start,
        end: exports.DateMapService.getInstance().yesterday_end,
    },
    "last_7_days": {
        start: exports.DateMapService.getInstance().day_7_start,
        end: exports.DateMapService.getInstance().today_end
    },
    "this_month": {
        start: exports.DateMapService.getInstance().month_start,
        end: exports.DateMapService.getInstance().today_end
    },
    "last_30_days": {
        start: exports.DateMapService.getInstance().day_30_start,
        end: exports.DateMapService.getInstance().today_end
    }
};
exports.DateRangeLabelMap = [
    {
        label: "Today",
        range: exports.DateRangeMap.today,
        isSingleDay: true,
        hasToday: true
    },
    {
        label: "Yesterday",
        range: exports.DateRangeMap.yesterday,
        isSingleDay: true,
        hasToday: false
    },
    {
        label: "Last 7 days",
        range: exports.DateRangeMap.last_7_days,
        isSingleDay: false,
        hasToday: true
    },
    {
        label: "This month",
        range: exports.DateRangeMap.this_month,
        isSingleDay: false,
        hasToday: true
    },
    {
        label: "Last 30 days",
        range: exports.DateRangeMap.last_30_days,
        isSingleDay: false,
        hasToday: true
    }
];
function isSameRange(range1, range2) {
    var start1 = range1.start;
    var start2 = range2.start;
    return date_fns_1.isSameDay(start1, start2) && date_fns_1.isSameDay(range1.end, range2.end);
}
exports.isSameRange = isSameRange;
function IsRangeToday(range) {
    if (range.start && range.end) {
        return range.start == date_fns_1.startOfToday().toISOString() && range.end == date_fns_1.endOfToday().toISOString();
    }
    else if (!range.start && !range.end) {
        return true;
    }
    else {
        return false;
    }
}
exports.IsRangeToday = IsRangeToday;
function GetDateRangeQuery(query, param) {
    if (param === void 0) { param = "created_at"; }
    var start = query["start"];
    var end = query["end"];
    return __assign({}, query, (_a = {}, _a["min_" + param] = start, _a["max_" + param] = end, _a.start = null, _a.end = null, _a));
    var _a;
}
exports.GetDateRangeQuery = GetDateRangeQuery;
function IsRangeADay(range) {
    if (range.start && range.end) {
        return date_fns_1.isSameDay(range.end, range.start);
        // return moment(range.end).diff(moment(range.start), "days") == 0;
    }
    else if (!range.start && !range.end) {
        return true;
    }
    else {
        return false;
    }
}
exports.IsRangeADay = IsRangeADay;
function RangeHasToday(query) {
    if (query.end) {
        return query.end == date_fns_1.endOfToday().toISOString();
    }
    else {
        return true;
    }
}
exports.RangeHasToday = RangeHasToday;
exports.dateRangeDisplay = function (range) {
    var rangeItem = exports.DateRangeLabelMap.find(function (item) {
        return exports.isSameDateRange(item.range, range);
    });
    if (rangeItem)
        return rangeItem.label;
    // const matchKey = Object.keys(DateRangeMap).find((key) => {
    //   return isSameDateRange(DateRangeMap[key], range)
    // });
    // return "";
    var customDate = exports.DateRangeLabelMap.find(function (data) {
        var customRange = data.range;
        return isSameRange(range, customRange);
    });
    if (customDate)
        return customDate.label;
    var isSingleDay = IsRangeADay(range);
    if (isSingleDay) {
        var isToday = IsRangeToday(range);
        var suffix = isToday ? "Today " : "";
        var string = suffix + date_string_1.DateString(range.start);
        return string;
    }
    else {
        // console.log(DateString(range.start), range.start);
        return date_string_1.DateString(range.start) + " - " + date_string_1.DateString(range.end);
    }
};
exports.isSameDateRange = function (range1, range2) {
    function nearTime(t1, t2) {
        return date_fns_1.isSameDay(t1, t2);
    }
    return (nearTime(range1.start, range2.start) && nearTime(range1.end, range2.end));
};
