// ht-client/dist/es/filters/actions-filter.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { BaseFilter } from "./base-filter";
import { Color } from "ht-utility";
var ActionsFilter = (function (_super) {
    __extends(ActionsFilter, _super);
    function ActionsFilter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.customQueryArray = [];
        _this.statusQueryArray = [
            {
                label: "Not yet started",
                values: ["created"],
                color: "#8a91a0"
            },
            {
                label: "Assigned",
                values: ["assigned", "started"],
                color: Color.blue
            },
            {
                label: "Completed",
                values: ["completed"],
                color: Color.stop
            },
            {
                label: "Suspended",
                values: ["suspended"],
                color: Color.red
            }
        ];
        _this.sortingQueryMap = {
            assigned_at: "Assigned",
            completed_at: "Completed",
            distance: "Distance",
            duration: "Duration",
            type: "Type",
            user__name: "User name",
            status: "Status"
        };
        _this.genericQueryArray = [];
        return _this;
    }
    Object.defineProperty(ActionsFilter.prototype, "allQueryArray", {
        get: function () {
            return this.statusQueryArray.concat(this.genericQueryArray, this.customQueryArray);
        },
        enumerable: true,
        configurable: true
    });
    return ActionsFilter;
}(BaseFilter));
export { ActionsFilter };
