// ht-client/dist/es/filters/users-filter.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { BaseFilter } from "./base-filter";
import { Color } from "ht-utility";
var DefaultUsersFilter = (function (_super) {
    __extends(DefaultUsersFilter, _super);
    function DefaultUsersFilter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.customQueryArray = [];
        _this.statusQueryArray = [
            {
                label: "Stopped",
                values: ["stopped"],
                color: Color.stop
            },
            {
                label: "Moving",
                values: ["on_trip"],
                color: Color.blue
            },
            {
                label: "Logged off",
                values: ["logged_off"],
                color: "#8a91a0"
            },
            {
                label: "Location disabled",
                values: ["location_disabled"],
                color: Color.red
            },
            {
                label: "Network offline",
                values: ["network_offline"],
                color: "#ccc"
            }
        ];
        _this.activityQueryArray = [
            {
                label: "Logged in",
                values: ["stopped", "on_trip", "network_offline"],
                color: Color.blue
            },
            {
                label: "Logged off",
                values: ["logged_off"],
                color: "#a8a8a8"
            },
            {
                label: "Location disabled",
                values: ["location_disabled"],
                color: Color.red
            }
        ];
        _this.genericQueryArray = [
            {
                label: "Show all",
                values: ["show_all"]
            }
        ];
        _this.showAllQueryArray = [
            {
                label: "Never tracked",
                values: ["never_tracked"],
                color: "#575757"
            }
        ];
        _this.sortingQueryMap = {
            num_trips: "Trips",
            name: "Name",
            status: "Status",
            num_places: "Places",
            total_distance: "Distance",
            total_duration: "Duration",
            stop_duration: "Stop duration",
            location_disabled_duration: "Location disabled",
            network_offline_duration: "Network offline",
            num_actions: "Actions",
            last_heartbeat_at: "Last updated"
        };
        _this.mapQueries = [];
        _this.statusOverviewQueries = ["search", "show_all"];
        return _this;
    }
    Object.defineProperty(DefaultUsersFilter.prototype, "allQueryArray", {
        get: function () {
            return this.statusQueryArray.concat(this.genericQueryArray, this.showAllQueryArray, this.customQueryArray);
        },
        enumerable: true,
        configurable: true
    });
    DefaultUsersFilter.prototype.getStatusQueryArray = function (showAll) {
        if (showAll === void 0) { showAll = false; }
        return showAll
            ? this.statusQueryArray.concat(this.showAllQueryArray) : this.statusQueryArray;
    };
    return DefaultUsersFilter;
}(BaseFilter));
export { DefaultUsersFilter };
