// ht-api/dist/es/entities/base.js
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
import { expand, map } from "rxjs/operators";
import { empty } from "rxjs/observable/empty";
var HtBaseApi = (function () {
    function HtBaseApi(request, base) {
        this.request = request;
        this.base = base;
    }
    HtBaseApi.prototype.get = function (id, query, token) {
        if (query === void 0) { query = {}; }
        var path = this.base + "/" + id + "/";
        return this.api$(path, query, { token: token });
    };
    HtBaseApi.prototype.index = function (query, token) {
        if (query === void 0) { query = {}; }
        var path = this.base + "/";
        return this.api$(path, query, { token: token });
    };
    HtBaseApi.prototype.summary = function (query, token) {
        if (query === void 0) { query = {}; }
        var path = this.base + "/summary/";
        return this.api$(path, query, { token: token });
    };
    HtBaseApi.prototype.heatmap = function (query, token) {
        if (query === void 0) { query = {}; }
        var path = this.base + "/heatmap/";
        return this.api$(path, query, { token: token });
    };
    HtBaseApi.prototype.api$ = function (path, query, options) {
        if (query === void 0) { query = {}; }
        if (options === void 0) { options = {}; }
        return this.request.api$(path, query, options);
    };
    HtBaseApi.prototype.postApi$ = function (path, body, options) {
        if (options === void 0) { options = {}; }
        return this.request.postApi$(path, body, options);
    };
    // getReqFromTail<T>(tail, query = {}, options = {}): Observable<T> {
    //   return this.request.api$(this.base + tail, query, options)
    // }
    //
    // postReqFromTail<T>(tail, body, options?): Observable<T> {
    //   return this.request.postApi$(this.base + tail, body, options)
    // }
    HtBaseApi.prototype.placeline = function (id, query, token) {
        if (query === void 0) { query = {}; }
        var tail = this.base + ("/" + id + "/placeline/");
        return this.api$(tail, query, { token: token });
    };
    HtBaseApi.prototype.allPages = function (api$, options) {
        var _this = this;
        if (options === void 0) { options = {}; }
        return api$.pipe(expand(function (data) {
            return data["next"]
                ? _this.request.api$(data["next"], {}, __assign({}, options, { pureUrl: true })).pipe(map(function (newData) {
                    return __assign({}, newData, { results: data.results.concat(newData.results) });
                }))
                : empty();
        }));
    };
    HtBaseApi.prototype.analytics = function (query) {
        return empty();
    };
    return HtBaseApi;
}());
export { HtBaseApi };
