// ht-maps/dist/lib/mixins2/style.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var StyleClass = (function () {
    function StyleClass() {
        this.defaultstyleFunct = {
            get: function (type) {
                return {
                    default: {}
                };
            }
        };
        this.highlightedId = null;
    }
    StyleClass.prototype.getStyle = function (selectedStyleType, fallbackStyle) {
        if (selectedStyleType === void 0) { selectedStyleType = "default"; }
        var mapType = this.mapInstance.mapUtils.type;
        var styleFunct = this.styleFunct || this.defaultstyleFunct;
        var mapTypetype = this.styleObj || styleFunct.get(mapType);
        // console.log(this.name, "style", selectedStyleType, styleFunct, this.styleFunct);
        // const styleType = mapTypetype[selectedStyleType] ? selectedStyleType : this.styleType;
        var style = mapTypetype[selectedStyleType] || fallbackStyle;
        if (!style)
            console.error("style type does not exist ", selectedStyleType);
        return style;
    };
    StyleClass.prototype.setStyleType = function (styleType) {
        if (styleType === void 0) { styleType = 'default'; }
        this.styleType = styleType;
    };
    StyleClass.prototype.setStyle = function (_a) {
        var item = _a.item, data = _a.data;
        var styleType = this.getStyleType(data);
        var style = this.getStyle(styleType);
        this.setItemStyle(item, style);
    };
    StyleClass.prototype.getStyleType = function (data) {
        if (this.highlightedId && data) {
            return this.highlightedId == data.id ? 'highlight' : 'fade';
        }
        else {
            return this.styleType;
        }
    };
    StyleClass.prototype.setItemStyle = function (item, style) {
        this.mapInstance.mapUtils.setStyle(item, style);
    };
    return StyleClass;
}());
exports.StyleClass = StyleClass;
