// ht-data/dist/es/rollup.config.js
import resolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import camelCase from 'lodash.camelcase';
var globals = [];
export var config = function (libraryName, pkg) {
    return {
        input: "dist/es/" + libraryName + ".js",
        output: [
            { file: pkg.main, name: camelCase(libraryName), format: 'umd' },
            { file: pkg.module, format: 'es' },
        ],
        sourcemap: true,
        // Indicate here external modules you don't wanna include in your bundle (i.e.: 'lodash')
        external: Object.keys(globals),
        globals: globals,
        watch: {
            include: 'dist/es/**',
        },
        plugins: [
            {
                name: 'replace moment imports',
                transform: function (code) {
                    return ({
                        code: code.replace(/import\s*\*\s*as\s*moment/g, 'import moment'),
                        map: { mappings: '' }
                    });
                }
            },
            // Allow bundling cjs modules (unlike webpack, rollup doesn't understand cjs)
            commonjs(),
            // Allow node_modules resolution, so you can use 'external' to control
            // which external modules to include in the bundle
            // https://github.com/rollup/rollup-plugin-node-resolve#usage
            resolve(),
        ],
    };
};
