// time-aware-polyline/dist/es/client/src/reducers/account-reducer.js
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
import * as AccountDispatcher from "../dispatchers/accounts-dispatcher";
import { createSelector } from "../store/selector";
export var initialState = {
    key: "false"
};
export function reducer(state, action) {
    if (state === void 0) { state = initialState; }
    switch (action.type) {
        case AccountDispatcher.SET_ACCOUNT_USER: {
            return __assign({}, state, { accountUser: action.payload });
        }
        case AccountDispatcher.SET_MEMBERHSHIPS_ALL: {
            // let results = state.membershipsAll ? [...state.membershipsAll.results, ...action.payload.results] : action.payload.results;
            // return {...state, membershipsAll: {...action.payload, results}}
            return __assign({}, state, { membershipsAll: action.payload });
        }
        case AccountDispatcher.SET_KEY: {
            return __assign({}, state, { key: action.payload });
        }
        case AccountDispatcher.SET_TEMP_KEY: {
            return __assign({}, state, { tempKey: action.payload });
        }
        case AccountDispatcher.SET_USER_ID: {
            return __assign({}, state, { userId: action.payload });
        }
        default: {
            return state;
        }
    }
}
export var getAccountUser = function (state) { return state.accountUser; };
export var getMembershipsAll = function (state) { return state.membershipsAll; };
export var getKey = function (state) { return state.key; };
export var getTempKey = function (state) { return state.tempKey; };
export var getUserId = function (state) { return state.userId; };
export var getCurrentKey = createSelector(getKey, getTempKey, function (key, tempKey) {
    return tempKey || key;
});
