// ht-data/dist/es/rx-helpers/data-intercepter.js
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
import { combineLatest } from "rxjs/observable/combineLatest";
import { distinctUntilChanged } from "rxjs/operators";
import { itemAsPage$ } from "./item-as-page";
export var dataWithSelectedId$ = function (data$, id$, keys, prop) {
    return combineLatest(data$, id$, function (data, id) {
        if (!data && keys.length)
            return data;
        var allItems = keys.reduce(function (items, key) {
            return items.concat(data[key]);
        }, []);
        var selected = allItems.find(function (item) {
            return item.id === id;
        });
        return __assign({}, data, (_a = {}, _a[prop] = selected, _a));
        var _a;
    });
};
export var listwithSelectedId$ = function (list$, id$) {
    return combineLatest(list$, id$.pipe(distinctUntilChanged()), function (list, id) {
        if (!list)
            return list;
        return !id
            ? list
            : {
                count: 1,
                next: null,
                page: null,
                results: list.results.filter(function (item) {
                    return item.id === id;
                })
            };
    });
};
export var listWithItem$ = function (list$, item$) {
    return combineLatest(list$, item$.pipe(itemAsPage$()), function (list, itemPage) {
        return itemPage ? itemPage : list;
    });
};
