// ht-data/dist/lib/entities/event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var HtEvent = (function () {
    function HtEvent(data) {
        this.data = data;
    }
    HtEvent.prototype.getEventDisplay = function (event) {
        event = event || this.data;
        switch (event.type) {
            case "tracking.started":
                return {
                    text: "Tracking started",
                    subText: ""
                };
            case "tracking.ended":
                return {
                    text: "Tracking ended",
                    subText: ""
                };
            // case 'device.location.disabled':
            //   return {
            //     text: 'Location disabled',
            //     subtext: ''
            //   };
            // case 'device.location.enabled':
            //   return {
            //     text: 'Location enabled',
            //     subtext: ''
            //   };
            // case 'device.location_permission.disabled':
            //   return {
            //     text: 'Location permission disabled',
            //     subtext: ''
            //   };
            // case 'device.location_permission.enabled':
            //   return {
            //     text: 'Location permission enabled',
            //     subtext: ''
            //   };
            case "device.secondary.ignored":
                return {
                    text: "Secondary device ignored",
                    subText: ""
                };
        }
    };
    return HtEvent;
}());
exports.HtEvent = HtEvent;
exports.htEvent = function (event) { return new HtEvent(event); };
