// ht-data/dist/es/utils/date-range.js
import { addDays, endOfToday, endOfYesterday, isSameDay, startOfDay, startOfMonth, startOfToday, startOfYesterday } from "date-fns";
import { DateString, IsRangeADay, IsRangeToday, isSameRange } from "ht-utility";
export var DateMapService = (function () {
    var instance;
    return {
        getInstance: function () {
            if (!instance) {
                instance = {
                    today_end: endOfToday().toISOString(),
                    today_start: startOfToday().toISOString(),
                    yesterday_start: startOfYesterday().toISOString(),
                    yesterday_end: endOfYesterday().toISOString(),
                    day_7_start: startOfDay(addDays(new Date(), -6)).toISOString(),
                    month_start: startOfMonth(new Date()).toISOString(),
                    day_30_start: startOfDay(addDays(new Date(), -29)).toISOString(),
                };
            }
            return instance;
        }
    };
})();
export var DateRangeMap = {
    "today": {
        start: DateMapService.getInstance().today_start,
        end: DateMapService.getInstance().today_end
    },
    "yesterday": {
        start: DateMapService.getInstance().yesterday_start,
        end: DateMapService.getInstance().yesterday_end,
    },
    "last_7_days": {
        start: DateMapService.getInstance().day_7_start,
        end: DateMapService.getInstance().today_end
    },
    "this_month": {
        start: DateMapService.getInstance().month_start,
        end: DateMapService.getInstance().today_end
    },
    "last_30_days": {
        start: DateMapService.getInstance().day_30_start,
        end: DateMapService.getInstance().today_end
    }
};
export var DateRangeLabelMap = [
    {
        label: "Today",
        range: DateRangeMap.today,
        isSingleDay: true,
        hasToday: true
    },
    {
        label: "Yesterday",
        range: DateRangeMap.yesterday,
        isSingleDay: true,
        hasToday: false
    },
    {
        label: "Last 7 days",
        range: DateRangeMap.last_7_days,
        isSingleDay: false,
        hasToday: true
    },
    {
        label: "This month",
        range: DateRangeMap.this_month,
        isSingleDay: false,
        hasToday: true
    },
    {
        label: "Last 30 days",
        range: DateRangeMap.last_30_days,
        isSingleDay: false,
        hasToday: true
    }
];
export var dateRangeDisplay = function (range) {
    var rangeItem = DateRangeLabelMap.find(function (item) {
        return isSameDateRange(item.range, range);
    });
    if (rangeItem)
        return rangeItem.label;
    // const matchKey = Object.keys(DateRangeMap).find((key) => {
    //   return isSameDateRange(DateRangeMap[key], range)
    // });
    // return "";
    var customDate = DateRangeLabelMap.find(function (data) {
        var customRange = data.range;
        return isSameRange(range, customRange);
    });
    if (customDate)
        return customDate.label;
    var isSingleDay = IsRangeADay(range);
    if (isSingleDay) {
        var isToday = IsRangeToday(range);
        var suffix = isToday ? "Today " : "";
        var string = suffix + DateString(range.start);
        return string;
    }
    else {
        // console.log(DateString(range.start), range.start);
        return DateString(range.start) + " - " + DateString(range.end);
    }
};
export var isSameDateRange = function (range1, range2) {
    function nearTime(t1, t2) {
        return isSameDay(t1, t2);
    }
    return (nearTime(range1.start, range2.start) && nearTime(range1.end, range2.end));
};
