// ht-client/dist/lib/dispatchers/accounts-dispatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SET_ACCOUNT_USER = "[ACCOUNT] set account user";
exports.SET_MEMBERHSHIPS_ALL = "[ACCOUNT] set memberships all";
exports.SET_KEY = "[ACCOUNT] set key";
exports.SET_TEMP_KEY = "[ACCOUNT] set temp key";
exports.SET_USER_ID = "[ACCOUNT] set user id";
var SetAccountUser = (function () {
    function SetAccountUser(payload) {
        this.payload = payload;
        this.type = exports.SET_ACCOUNT_USER;
    }
    return SetAccountUser;
}());
exports.SetAccountUser = SetAccountUser;
var SetMembershipsAll = (function () {
    function SetMembershipsAll(payload) {
        this.payload = payload;
        this.type = exports.SET_MEMBERHSHIPS_ALL;
    }
    return SetMembershipsAll;
}());
exports.SetMembershipsAll = SetMembershipsAll;
var SetKey = (function () {
    function SetKey(payload) {
        this.payload = payload;
        this.type = exports.SET_KEY;
    }
    return SetKey;
}());
exports.SetKey = SetKey;
var SetTempKey = (function () {
    function SetTempKey(payload) {
        this.payload = payload;
        this.type = exports.SET_TEMP_KEY;
    }
    return SetTempKey;
}());
exports.SetTempKey = SetTempKey;
var SetUserId = (function () {
    function SetUserId(payload) {
        this.payload = payload;
        this.type = exports.SET_USER_ID;
    }
    return SetUserId;
}());
exports.SetUserId = SetUserId;
