// ht-client/dist/lib/dispatchers/actions-dispatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
List
 */
exports.SET_ACTIONS_LIST = "[ACTIONS] set actions list";
exports.SET_ACTIONS_LIST_QUERY = "[ACTIONS] set actions list query";
exports.ADD_ACTIONS_LIST_QUERY = "[ACTIONS] add actions list query";
exports.SET_ACTIONS_LIST_ACTIVE = "[ACTIONS] set actions list active";
exports.SET_ACTIONS_LIST_LOADING = "[ACTIONS] set actions list loading";
/*
Graph
 */
exports.SET_ACTIONS_GRAPH = "[ACTIONS] set actions graph";
exports.SET_ACTIONS_GRAPH_QUERY = "[ACTIONS] set actions graph query";
exports.ADD_ACTIONS_GRAPH_QUERY = "[ACTIONS] add actions graph query";
exports.SET_ACTIONS_GRAPH_LOADING = "[ACTIONS] set actions graph loading";
/*
Summary
 */
exports.SET_ACTIONS_SUMMARY = "[ACTIONS] set actions summary";
exports.SET_ACTIONS_SUMMARY_ACTIVE = "[ACTIONS] set actions summary active";
exports.ADD_ACTIONS_SUMMARY_QUERY = "[ACTIONS] actions summary query";
exports.SET_ACTIONS_SUMMARY_QUERY = "[ACTIONS] add set actions summary query";
exports.SET_ACTIONS_SUMMARY_LOADING = "[ACTIONS] set actions summary loading";
/*
Heatmap
 */
// export const SET_ACTIONS_HEATMAP = "[ACTIONS] set actions summary";
// export const SET_ACTIONS_HEATMAP_ACTIVE = "[ACTIONS] set actions summary active";
// export const ADD_ACTIONS_HEATMAP_QUERY = "[ACTIONS] actions summary query";
// export const SET_ACTIONS_HEATMAP_QUERY = "[ACTIONS] add set actions summary query";
// export const SET_ACTIONS_HEATMAP_LOADING = "[ACTIONS] set actions graph loading";
var SetList = (function () {
    function SetList(payload) {
        this.payload = payload;
        this.type = exports.SET_ACTIONS_LIST;
    }
    return SetList;
}());
exports.SetList = SetList;
var SetListQuery = (function () {
    function SetListQuery(payload) {
        this.payload = payload;
        this.type = exports.SET_ACTIONS_LIST_QUERY;
    }
    return SetListQuery;
}());
exports.SetListQuery = SetListQuery;
var AddListQuery = (function () {
    function AddListQuery(payload) {
        this.payload = payload;
        this.type = exports.ADD_ACTIONS_LIST_QUERY;
    }
    return AddListQuery;
}());
exports.AddListQuery = AddListQuery;
var SetListActive = (function () {
    function SetListActive(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = exports.SET_ACTIONS_LIST_ACTIVE;
    }
    return SetListActive;
}());
exports.SetListActive = SetListActive;
var SetListLoading = (function () {
    function SetListLoading(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = exports.SET_ACTIONS_LIST_LOADING;
    }
    return SetListLoading;
}());
exports.SetListLoading = SetListLoading;
var SetGraph = (function () {
    function SetGraph(payload) {
        this.payload = payload;
        this.type = exports.SET_ACTIONS_GRAPH;
    }
    return SetGraph;
}());
exports.SetGraph = SetGraph;
var SetGraphQuery = (function () {
    function SetGraphQuery(payload) {
        this.payload = payload;
        this.type = exports.SET_ACTIONS_GRAPH_QUERY;
    }
    return SetGraphQuery;
}());
exports.SetGraphQuery = SetGraphQuery;
var SetGraphLoading = (function () {
    function SetGraphLoading(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = exports.SET_ACTIONS_GRAPH_LOADING;
    }
    return SetGraphLoading;
}());
exports.SetGraphLoading = SetGraphLoading;
var AddGraphQuery = (function () {
    function AddGraphQuery(payload) {
        this.payload = payload;
        this.type = exports.ADD_ACTIONS_GRAPH_QUERY;
    }
    return AddGraphQuery;
}());
exports.AddGraphQuery = AddGraphQuery;
var SetSummary = (function () {
    function SetSummary(payload) {
        this.payload = payload;
        this.type = exports.SET_ACTIONS_SUMMARY;
    }
    return SetSummary;
}());
exports.SetSummary = SetSummary;
var AddSummaryQuery = (function () {
    function AddSummaryQuery(payload) {
        this.payload = payload;
        this.type = exports.ADD_ACTIONS_SUMMARY_QUERY;
    }
    return AddSummaryQuery;
}());
exports.AddSummaryQuery = AddSummaryQuery;
var SetSummaryQuery = (function () {
    function SetSummaryQuery(payload) {
        this.payload = payload;
        this.type = exports.SET_ACTIONS_SUMMARY_QUERY;
    }
    return SetSummaryQuery;
}());
exports.SetSummaryQuery = SetSummaryQuery;
var SetSummaryLoading = (function () {
    function SetSummaryLoading(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = exports.SET_ACTIONS_SUMMARY_LOADING;
    }
    return SetSummaryLoading;
}());
exports.SetSummaryLoading = SetSummaryLoading;
var SetSummaryActive = (function () {
    function SetSummaryActive(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = exports.SET_ACTIONS_SUMMARY_ACTIVE;
    }
    return SetSummaryActive;
}());
exports.SetSummaryActive = SetSummaryActive;
