// ht-client/dist/lib/dispatchers/groups-dispatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SET_LIST_ACTIVE = "[GROUPS] set list active";
exports.SET_ID = "[GROUPS] set id";
exports.SET_GROUPS = "[GROUPS] set groups";
exports.SET_GROUP = "[GROUPS] set group";
var SetListActive = (function () {
    function SetListActive(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = exports.SET_LIST_ACTIVE;
    }
    return SetListActive;
}());
exports.SetListActive = SetListActive;
var SetGroupId = (function () {
    function SetGroupId(payload) {
        this.payload = payload;
        this.type = exports.SET_ID;
    }
    return SetGroupId;
}());
exports.SetGroupId = SetGroupId;
var SetGroupsAll = (function () {
    function SetGroupsAll(payload) {
        this.payload = payload;
        this.type = exports.SET_GROUPS;
    }
    return SetGroupsAll;
}());
exports.SetGroupsAll = SetGroupsAll;
var SetGroup = (function () {
    function SetGroup(payload) {
        this.payload = payload;
        this.type = exports.SET_GROUP;
    }
    return SetGroup;
}());
exports.SetGroup = SetGroup;
