// ht-client/dist/lib/dispatchers/loading-dispatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SET_USER_ANALYTICS = "[LOADING] set user analytics";
exports.SET_USER_INDEX = "[LOADING] set user index";
exports.SET_USER_ANALYTICS_ALL = "[LOADING] set user analytics all";
exports.SET_USER_INDEX_ALL = "[LOADING] set user index all";
var SetLoadingUserAnalytics = (function () {
    function SetLoadingUserAnalytics(payload) {
        this.payload = payload;
        this.type = exports.SET_USER_ANALYTICS;
    }
    return SetLoadingUserAnalytics;
}());
exports.SetLoadingUserAnalytics = SetLoadingUserAnalytics;
var SetLoadingUserIndex = (function () {
    function SetLoadingUserIndex(payload) {
        this.payload = payload;
        this.type = exports.SET_USER_INDEX;
    }
    return SetLoadingUserIndex;
}());
exports.SetLoadingUserIndex = SetLoadingUserIndex;
var SetLoadingUserAnalyticsAll = (function () {
    function SetLoadingUserAnalyticsAll(payload) {
        this.payload = payload;
        this.type = exports.SET_USER_ANALYTICS_ALL;
    }
    return SetLoadingUserAnalyticsAll;
}());
exports.SetLoadingUserAnalyticsAll = SetLoadingUserAnalyticsAll;
var SetLoadingUserIndexAll = (function () {
    function SetLoadingUserIndexAll(payload) {
        this.payload = payload;
        this.type = exports.SET_USER_INDEX_ALL;
    }
    return SetLoadingUserIndexAll;
}());
exports.SetLoadingUserIndexAll = SetLoadingUserIndexAll;
