// ht-client/dist/lib/dispatchers/query-dispatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SET_DATE_RANGE = "[QUERY] set date range";
exports.SET_USER_QUERY = "[QUERY] set user query";
exports.SET_USER_QUERY_RESET_PAGE = "[QUERY] set user query reset page";
exports.CLEAR_USER_QUERY_KEY = "[QUERY] clear user query key";
exports.SET_USER_ID = "[USERS] set user id";
exports.TOGGLE_USER_ID = "[USERS] toggle user id";
exports.SET_PLACELINE_QUERY = "[USERS] Set placeline query";
var SetDateRange = (function () {
    function SetDateRange(payload) {
        this.payload = payload;
        this.type = exports.SET_DATE_RANGE;
    }
    return SetDateRange;
}());
exports.SetDateRange = SetDateRange;
var SetUserQuery = (function () {
    function SetUserQuery(payload) {
        this.payload = payload;
        this.type = exports.SET_USER_QUERY;
    }
    return SetUserQuery;
}());
exports.SetUserQuery = SetUserQuery;
var SetUserQueryResetPage = (function () {
    function SetUserQueryResetPage(payload) {
        this.payload = payload;
        this.type = exports.SET_USER_QUERY_RESET_PAGE;
    }
    return SetUserQueryResetPage;
}());
exports.SetUserQueryResetPage = SetUserQueryResetPage;
var ClearUserQueryKey = (function () {
    function ClearUserQueryKey(payload) {
        this.payload = payload;
        this.type = exports.CLEAR_USER_QUERY_KEY;
    }
    return ClearUserQueryKey;
}());
exports.ClearUserQueryKey = ClearUserQueryKey;
var SetUserId = (function () {
    function SetUserId(payload) {
        this.payload = payload;
        this.type = exports.SET_USER_ID;
    }
    return SetUserId;
}());
exports.SetUserId = SetUserId;
var ToggleUserId = (function () {
    function ToggleUserId(payload) {
        this.payload = payload;
        this.type = exports.TOGGLE_USER_ID;
    }
    return ToggleUserId;
}());
exports.ToggleUserId = ToggleUserId;
var SetPlacelineQuery = (function () {
    function SetPlacelineQuery(payload) {
        if (payload === void 0) { payload = {}; }
        this.payload = payload;
        this.type = exports.SET_PLACELINE_QUERY;
    }
    return SetPlacelineQuery;
}());
exports.SetPlacelineQuery = SetPlacelineQuery;
