// ht-client/dist/lib/dispatchers/segments-dispatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SET_SELECTED_ID = "[SEGMENTS] set selected id";
exports.SET_HIGHLIGHTED_ID = "[SEGMENTS] set highlighted id";
exports.SET_RESET_MAP_ID = "[SEGMENTS] set reset map id";
exports.SET_POPUP_ID = "[SEGMENTS] set popup id";
var SetSelectedId = (function () {
    function SetSelectedId(payload) {
        this.payload = payload;
        this.type = exports.SET_SELECTED_ID;
    }
    return SetSelectedId;
}());
exports.SetSelectedId = SetSelectedId;
var SetHighlightedId = (function () {
    function SetHighlightedId(payload) {
        this.payload = payload;
        this.type = exports.SET_HIGHLIGHTED_ID;
    }
    return SetHighlightedId;
}());
exports.SetHighlightedId = SetHighlightedId;
var SetResetMapId = (function () {
    function SetResetMapId(payload) {
        this.payload = payload;
        this.type = exports.SET_RESET_MAP_ID;
    }
    return SetResetMapId;
}());
exports.SetResetMapId = SetResetMapId;
var SetPopupId = (function () {
    function SetPopupId(payload) {
        this.payload = payload;
        this.type = exports.SET_POPUP_ID;
    }
    return SetPopupId;
}());
exports.SetPopupId = SetPopupId;
