// ht-client/dist/lib/dispatchers/user-dispatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
//placeline
exports.SET_USER_DATA = "[USERS] set user data";
exports.SET_PLACELINE_ID = "[USERS] set placeline id";
exports.TOGGLE_PLACELINE_ID = "[USERS] toggle placeline id";
exports.SET_PLACELINE_QUERY = "[USERS] set placeline query";
exports.ADD_PLACELINE_QUERY = "[USERS] add placeline query";
exports.SET_PLACELINE_LOADING = "[USERS] set placeline loading";
//list
exports.SET_LIST_ACTIVE = "[USERS] set list active";
exports.SET_LIST_ID = "[USERS] set list id";
exports.SET_LIST_QUERY = "[USERS] set list query";
exports.ADD_LIST_QUERY = "[USERS] add list query";
exports.CLEAR_QUERY_KEY = "[USERS] clear query key";
exports.TOGGLE_LIST_ID = "[USERS] toggle list id";
//analytics page
exports.SET_USERS_ANALYTICS_PAGE = "[USERS] set user analytics page";
exports.SET_USERS_ANALYTICS_ID = "[USERS] set analytics id";
exports.SET_USERS_ANALYTICS_QUERY = "[USERS] set analytics query";
exports.SET_USERS_ANALYTICS_LOADING = "[USERS] set analytics loading";
exports.SET_USERS_LIST_DATA_MAP = "[USERS] set users list data map";
//analytics all
exports.SET_MARKERS_ACTIVE = "[USERS] set markerS active";
exports.SET_USERS_ANALYTICS_ALL = "[USERS] set users analytics all";
exports.ADD_USERS_ANALYTICS_ALL = "[USERS] add users analytics all";
exports.SET_USERS_ANALYTICS_ALL_QUERY = "[USERS] add analytics all query";
exports.ADD_USERS_ANALYTICS_ALL_QUERY = "[USERS] set analytics all query";
exports.SET_USERS_ANALYTICS_ALL_LOADING = "[USERS] set analytics all loading";
exports.SET_USERS_MARKERS_DATA_MAP = "[USERS] set users listAll data map";
//summary
exports.SET_SUMMARY_ACTIVE = "[USERS] set summary active";
exports.SET_USERS_SUMMARY = "[USERS] set users summary";
exports.SET_USERS_SUMMARY_LOADING = "[USERS] set summary loading";
exports.SET_USERS_INDEX_PAGE = "[USERS] set user index page";
exports.ADD_USERS_INDEX_ALL = "[USERS] add users index all";
exports.SET_USERS_LIST_API_TYPE = "[USERS] set users list api type";
var SetUserData = (function () {
    function SetUserData(payload) {
        this.payload = payload;
        this.type = exports.SET_USER_DATA;
    }
    return SetUserData;
}());
exports.SetUserData = SetUserData;
var SetPlacelineId = (function () {
    function SetPlacelineId(payload) {
        this.payload = payload;
        this.type = exports.SET_PLACELINE_ID;
    }
    return SetPlacelineId;
}());
exports.SetPlacelineId = SetPlacelineId;
var TogglePlacelineId = (function () {
    function TogglePlacelineId(payload) {
        this.payload = payload;
        this.type = exports.TOGGLE_PLACELINE_ID;
    }
    return TogglePlacelineId;
}());
exports.TogglePlacelineId = TogglePlacelineId;
var SetPlacelineQuery = (function () {
    function SetPlacelineQuery(payload) {
        this.payload = payload;
        this.type = exports.SET_PLACELINE_QUERY;
    }
    return SetPlacelineQuery;
}());
exports.SetPlacelineQuery = SetPlacelineQuery;
var AddPlacelineQuery = (function () {
    function AddPlacelineQuery(payload) {
        this.payload = payload;
        this.type = exports.ADD_PLACELINE_QUERY;
    }
    return AddPlacelineQuery;
}());
exports.AddPlacelineQuery = AddPlacelineQuery;
var SetPlacelineLoading = (function () {
    function SetPlacelineLoading(payload) {
        this.payload = payload;
        this.type = exports.SET_PLACELINE_LOADING;
    }
    return SetPlacelineLoading;
}());
exports.SetPlacelineLoading = SetPlacelineLoading;
var SetUsersListId = (function () {
    function SetUsersListId(payload) {
        this.payload = payload;
        this.type = exports.SET_LIST_ID;
    }
    return SetUsersListId;
}());
exports.SetUsersListId = SetUsersListId;
var ToggleUsersListId = (function () {
    function ToggleUsersListId(payload) {
        this.payload = payload;
        this.type = exports.TOGGLE_LIST_ID;
    }
    return ToggleUsersListId;
}());
exports.ToggleUsersListId = ToggleUsersListId;
var SetListQuery = (function () {
    function SetListQuery(payload) {
        this.payload = payload;
        this.type = exports.SET_LIST_QUERY;
    }
    return SetListQuery;
}());
exports.SetListQuery = SetListQuery;
var AddListQuery = (function () {
    function AddListQuery(payload) {
        this.payload = payload;
        this.type = exports.ADD_LIST_QUERY;
    }
    return AddListQuery;
}());
exports.AddListQuery = AddListQuery;
var ClearQueryKey = (function () {
    function ClearQueryKey(payload) {
        this.payload = payload;
        this.type = exports.CLEAR_QUERY_KEY;
    }
    return ClearQueryKey;
}());
exports.ClearQueryKey = ClearQueryKey;
var SetUsersAnalyticsPage = (function () {
    function SetUsersAnalyticsPage(payload) {
        this.payload = payload;
        this.type = exports.SET_USERS_ANALYTICS_PAGE;
    }
    return SetUsersAnalyticsPage;
}());
exports.SetUsersAnalyticsPage = SetUsersAnalyticsPage;
var SetUsersAnalyticsLoading = (function () {
    function SetUsersAnalyticsLoading(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = exports.SET_USERS_ANALYTICS_LOADING;
    }
    return SetUsersAnalyticsLoading;
}());
exports.SetUsersAnalyticsLoading = SetUsersAnalyticsLoading;
var SetUsersIndexPage = (function () {
    function SetUsersIndexPage(payload) {
        this.payload = payload;
        this.type = exports.SET_USERS_INDEX_PAGE;
    }
    return SetUsersIndexPage;
}());
exports.SetUsersIndexPage = SetUsersIndexPage;
var AddUsersAnalyticsAll = (function () {
    function AddUsersAnalyticsAll(payload) {
        this.payload = payload;
        this.type = exports.ADD_USERS_ANALYTICS_ALL;
    }
    return AddUsersAnalyticsAll;
}());
exports.AddUsersAnalyticsAll = AddUsersAnalyticsAll;
var SetUsersAnalyticsAll = (function () {
    function SetUsersAnalyticsAll(payload) {
        this.payload = payload;
        this.type = exports.SET_USERS_ANALYTICS_ALL;
    }
    return SetUsersAnalyticsAll;
}());
exports.SetUsersAnalyticsAll = SetUsersAnalyticsAll;
var SetUsersAnalyticsAllLoading = (function () {
    function SetUsersAnalyticsAllLoading(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = exports.SET_USERS_ANALYTICS_ALL_LOADING;
    }
    return SetUsersAnalyticsAllLoading;
}());
exports.SetUsersAnalyticsAllLoading = SetUsersAnalyticsAllLoading;
var SetUsersIndexAll = (function () {
    function SetUsersIndexAll(payload) {
        this.payload = payload;
        //todo change name
        this.type = exports.ADD_USERS_INDEX_ALL;
    }
    return SetUsersIndexAll;
}());
exports.SetUsersIndexAll = SetUsersIndexAll;
var SetUsersListApiType = (function () {
    function SetUsersListApiType(payload) {
        this.payload = payload;
        this.type = exports.SET_USERS_LIST_API_TYPE;
    }
    return SetUsersListApiType;
}());
exports.SetUsersListApiType = SetUsersListApiType;
var SetUsersListDataMap = (function () {
    function SetUsersListDataMap(payload) {
        this.payload = payload;
        this.type = exports.SET_USERS_LIST_DATA_MAP;
    }
    return SetUsersListDataMap;
}());
exports.SetUsersListDataMap = SetUsersListDataMap;
var SetUsersMarkersDataMap = (function () {
    function SetUsersMarkersDataMap(payload) {
        this.payload = payload;
        this.type = exports.SET_USERS_MARKERS_DATA_MAP;
    }
    return SetUsersMarkersDataMap;
}());
exports.SetUsersMarkersDataMap = SetUsersMarkersDataMap;
var SetListActive = (function () {
    function SetListActive(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = exports.SET_LIST_ACTIVE;
    }
    return SetListActive;
}());
exports.SetListActive = SetListActive;
var SetMarkersActive = (function () {
    function SetMarkersActive(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = exports.SET_MARKERS_ACTIVE;
    }
    return SetMarkersActive;
}());
exports.SetMarkersActive = SetMarkersActive;
var SetSummaryActive = (function () {
    function SetSummaryActive(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = exports.SET_SUMMARY_ACTIVE;
    }
    return SetSummaryActive;
}());
exports.SetSummaryActive = SetSummaryActive;
var SetUsersSummary = (function () {
    function SetUsersSummary(payload) {
        this.payload = payload;
        this.type = exports.SET_USERS_SUMMARY;
    }
    return SetUsersSummary;
}());
exports.SetUsersSummary = SetUsersSummary;
var SetSummaryLoading = (function () {
    function SetSummaryLoading(payload) {
        if (payload === void 0) { payload = true; }
        this.payload = payload;
        this.type = exports.SET_USERS_SUMMARY_LOADING;
    }
    return SetSummaryLoading;
}());
exports.SetSummaryLoading = SetSummaryLoading;
