// ht-maps/dist/lib/compound-entities/action-trace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var destination_marker_1 = require("../entities/destination-marker");
var actions_data_polyline_1 = require("../entities/actions-data-polyline");
var start_marker_1 = require("../entities/start-marker");
var operators_1 = require("rxjs/operators");
var action_user_1 = require("../entities/action-user");
var animations_entities_1 = require("../mixins/animations-entities");
var ActionMap = (function () {
    function ActionMap(mapInstance, options) {
        if (options === void 0) { options = {}; }
        this.anim = new animations_entities_1.AnimationsEntities();
        this.mapInstance = mapInstance;
        this.destination = new destination_marker_1.DestinationMarker(mapInstance);
        this.polyline = new actions_data_polyline_1.ActionsPolylineTrace(mapInstance);
        this.polyline.setTimeAwareAnimationEntity(this.anim);
        this.user = new action_user_1.ActionUserTrace(mapInstance);
        this.user.setTimeAwareAnimationEntity(this.anim);
        if (options.hasPulse) {
            this.pulse = new action_user_1.ActionUserTrace(mapInstance);
            this.pulse.setTimeAwareAnimationEntity(this.anim);
        }
        // this.user.setTimeAwareAnimation(this.anim);
        // this.user.toNotTraceItem = true;
        this.start = new start_marker_1.StartMarkerTrace(mapInstance);
    }
    ActionMap.prototype.setData$ = function (data$) {
        this.destination.setData$(data$);
        this.polyline.setData$(data$);
        this.start.setData$(data$);
        if (this.pulse)
            this.pulse.setData$(data$.pipe(operators_1.map(function (actions) {
                return actions.reduce(function (acc, action) {
                    return action.display.show_summary ? acc : acc.concat([action]);
                }, []);
            })));
        this.user.setData$(data$.pipe(operators_1.map(function (actions) {
            return actions.reduce(function (acc, action) {
                return action.display.show_summary ? acc : acc.concat([action]);
            }, []);
        })));
    };
    ;
    return ActionMap;
}());
exports.ActionMap = ActionMap;
exports.ActionTrace = ActionMap;
