// ht-maps/dist/lib/mixins/marker-renderer.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("underscore");
var ht_utility_1 = require("ht-utility");
function MarkersMixin(Base) {
    return (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.entities = {};
            _this.event = new ht_utility_1.HtCustomEvent();
            return _this;
        }
        class_1.prototype.htShow = function (item) {
            return "display: " + (item ? "flex" : "none");
        };
        class_1.prototype.getItem = function (data) {
            return this.mapInstance.mapUtils.getMarker();
        };
        class_1.prototype.getBounds = function (item, bounds) {
            return this.mapInstance.mapUtils.extendItemBounds(item, bounds, this.forceExtendBounds);
        };
        class_1.prototype.update = function (_a, positionBearing) {
            var item = _a.item, data = _a.data;
            var pathPosition = positionBearing && positionBearing.path.length ?
                positionBearing.path[positionBearing.path.length - 1] : null;
            var position = pathPosition || this.getPosition(data);
            if (position)
                this.mapInstance.mapUtils.updatePosition(item, position);
            if (!this.toNotSetMap)
                this.mapInstance.mapUtils.setMap(item, this.mapInstance.map);
            var id = this.trackBy(data);
            this.event.next('update-' + id, { item: item, data: data });
        };
        /**
         *
         * @param id string return trackBy
         * @returns {{subscribe: (cb) => IEventSub}}
         */
        class_1.prototype.onEntityUpdate = function (id) {
            var _this = this;
            var eventName = "update-" + id;
            return {
                subscribe: function (cb) {
                    return _this.event.subscribe(eventName, cb);
                }
            };
        };
        class_1.prototype.removeItem = function (item) {
            this.mapInstance.mapUtils.clearItem(item);
        };
        class_1.prototype.removeAll = function (entities) {
            var _this = this;
            _.each(entities, function (entity) {
                _this.removeItem(entity.item);
            });
            this.entities = {};
        };
        class_1.prototype.clear = function () {
            var entities = this.entities;
            this.removeAll(entities);
        };
        class_1.prototype.clearItem = function (_a) {
            var item = _a.item, data = _a.data;
            this.removeData(data);
            this.removeItem(item);
        };
        class_1.prototype.removeData = function (data) {
            var id = this.trackBy(data);
            if (this.entities[id])
                delete this.entities[id];
        };
        return class_1;
    }(Base));
}
exports.MarkersMixin = MarkersMixin;
