// ht-maps/dist/lib/styles/segment-polyline-styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ht_utility_1 = require("ht-utility");
exports.segmentPolylineStyles = {
    get: function (type) {
        switch (type) {
            case "google":
                {
                    return {
                        default: {
                            strokeColor: ht_utility_1.Color.blue,
                            strokeOpacity: 1,
                            strokeWeight: 5
                        },
                        highlight: {
                            strokeColor: ht_utility_1.Color.blue,
                            strokeOpacity: 1,
                            strokeWeight: 5
                        },
                        fade: {
                            strokeColor: ht_utility_1.Color.grey5,
                            strokeOpacity: 0.2,
                            strokeWeight: 2
                        }
                    };
                }
                ;
            case "leaflet": {
                return {
                    default: {
                        weight: 5,
                        color: ht_utility_1.Color.blue,
                        opacity: 1
                    },
                    highlight: {
                        weight: 5,
                        color: ht_utility_1.Color.blue,
                        opacity: 1
                    },
                    fade: {
                        weight: 2,
                        color: ht_utility_1.Color.grey2,
                        opacity: 0.2
                    }
                };
            }
        }
    },
};
