// ht-data/dist/lib/utils/date-range.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var date_fns_1 = require("date-fns");
var ht_utility_1 = require("ht-utility");
exports.DateMapService = (function () {
    var instance;
    return {
        getInstance: function () {
            if (!instance) {
                instance = {
                    today_end: date_fns_1.endOfToday().toISOString(),
                    today_start: date_fns_1.startOfToday().toISOString(),
                    yesterday_start: date_fns_1.startOfYesterday().toISOString(),
                    yesterday_end: date_fns_1.endOfYesterday().toISOString(),
                    day_7_start: date_fns_1.startOfDay(date_fns_1.addDays(new Date(), -6)).toISOString(),
                    month_start: date_fns_1.startOfMonth(new Date()).toISOString(),
                    day_30_start: date_fns_1.startOfDay(date_fns_1.addDays(new Date(), -29)).toISOString(),
                };
            }
            return instance;
        }
    };
})();
exports.DateRangeMap = {
    "today": {
        start: exports.DateMapService.getInstance().today_start,
        end: exports.DateMapService.getInstance().today_end
    },
    "yesterday": {
        start: exports.DateMapService.getInstance().yesterday_start,
        end: exports.DateMapService.getInstance().yesterday_end,
    },
    "last_7_days": {
        start: exports.DateMapService.getInstance().day_7_start,
        end: exports.DateMapService.getInstance().today_end
    },
    "this_month": {
        start: exports.DateMapService.getInstance().month_start,
        end: exports.DateMapService.getInstance().today_end
    },
    "last_30_days": {
        start: exports.DateMapService.getInstance().day_30_start,
        end: exports.DateMapService.getInstance().today_end
    }
};
exports.DateRangeLabelMap = [
    {
        label: "Today",
        range: exports.DateRangeMap.today,
        isSingleDay: true,
        hasToday: true
    },
    {
        label: "Yesterday",
        range: exports.DateRangeMap.yesterday,
        isSingleDay: true,
        hasToday: false
    },
    {
        label: "Last 7 days",
        range: exports.DateRangeMap.last_7_days,
        isSingleDay: false,
        hasToday: true
    },
    {
        label: "This month",
        range: exports.DateRangeMap.this_month,
        isSingleDay: false,
        hasToday: true
    },
    {
        label: "Last 30 days",
        range: exports.DateRangeMap.last_30_days,
        isSingleDay: false,
        hasToday: true
    }
];
exports.dateRangeDisplay = function (range) {
    var rangeItem = exports.DateRangeLabelMap.find(function (item) {
        return exports.isSameDateRange(item.range, range);
    });
    if (rangeItem)
        return rangeItem.label;
    // const matchKey = Object.keys(DateRangeMap).find((key) => {
    //   return isSameDateRange(DateRangeMap[key], range)
    // });
    // return "";
    var customDate = exports.DateRangeLabelMap.find(function (data) {
        var customRange = data.range;
        return ht_utility_1.isSameRange(range, customRange);
    });
    if (customDate)
        return customDate.label;
    var isSingleDay = ht_utility_1.IsRangeADay(range);
    if (isSingleDay) {
        var isToday = ht_utility_1.IsRangeToday(range);
        var suffix = isToday ? "Today " : "";
        var string = suffix + ht_utility_1.DateString(range.start);
        return string;
    }
    else {
        // console.log(DateString(range.start), range.start);
        return ht_utility_1.DateString(range.start) + " - " + ht_utility_1.DateString(range.end);
    }
};
exports.isSameDateRange = function (range1, range2) {
    function nearTime(t1, t2) {
        return date_fns_1.isSameDay(t1, t2);
    }
    return (nearTime(range1.start, range2.start) && nearTime(range1.end, range2.end));
};
