// ht-maps/dist/lib/entities/actions-cluster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ht_data_1 = require("ht-data");
var ht_utility_1 = require("ht-utility");
var data_observable_1 = require("../mixins/data-observable");
var popup_renderer_1 = require("../mixins/popup-renderer");
var clusters_1 = require("../mixins/clusters");
var extend_bounds_1 = require("../mixins/extend-bounds");
var marker_renderer_1 = require("../mixins/marker-renderer");
var styles_1 = require("../mixins/styles");
var trace_1 = require("../mixins/trace");
var div_markers_renderes_1 = require("../mixins/div-markers-renderes");
var map_items_1 = require("../mixins/map-items");
var ActionsCluster = (function () {
    function ActionsCluster(mapInstance) {
        this.mapInstance = mapInstance;
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case 'google': {
                        return {
                            default: {
                                flat: true,
                                anchor: RichMarkerPosition.MIDDLE,
                                zIndex: 1
                            },
                            popup: {
                                pixelOffset: new google.maps.Size(0, -5),
                            }
                        };
                    }
                    case 'leaflet': {
                        return {
                            default: {
                                iconAnchor: [12, 12]
                                // iconSize: [35, 35],
                                // className: 'current-action-marker',
                                // iconAnchor: point(15, 43)
                                // iconAnchor: [15, 43]
                            },
                            popup: {
                                // offset: point(0, -35),
                                offset: [0, -5],
                                closeButton: false
                            }
                        };
                    }
                }
            }
        };
    }
    ActionsCluster.prototype.getPosition = function (data) {
        return ht_data_1.htAction(data).getPosition();
    };
    ActionsCluster.prototype.getDivContent = function (action) {
        var icon = "<div class=\"action-marker flex-row\">\n<span style=\"margin: auto\">" + ht_utility_1.NameCase(action.type[0]) + "</span>\n</div>";
        return icon;
    };
    ActionsCluster.prototype.getInfoContent = function (item) {
        function htShow(item, style) {
            if (style === void 0) { style = "flex"; }
            return "display: " + (item ? style : "none");
        }
        var userName = item.user ? item.user.name : '';
        return "<div class=\"flex-column flex-center\" style=\"min-width: 180px\">\n<div class=\"\">\n    <div class=\"text-center\">" + ht_utility_1.NameCase(item.type) + "\n    <span style=\"" + htShow(item.display.duration_remaining && !item.display.show_summary) + "\"> in " + ht_utility_1.HMString(item.display.duration_remaining / 60) + "</span>\n    <span style=\"" + htShow(!!item.completed_at) + "\"> completed at " + ht_utility_1.TimeString(item.completed_at) + "</span>\n   \n    </div>\n</div>\n    <div class=\"text-muted text-center\" style=\"" + htShow(!!item.completed_at) + "\"> " + ht_utility_1.DateString(item.completed_at) + "</div>\n<div class=\"text-center\">" + ht_utility_1.NameCase(userName) + "<span style=\"" + htShow(!!item.lookup_id, 'block') + "\"> | #" + item.lookup_id + "</span></div>\n</div>";
    };
    ActionsCluster.prototype.traceEffect = function () {
    };
    return ActionsCluster;
}());
exports.ActionsCluster = ActionsCluster;
;
exports.ActionsClusterTrace = data_observable_1.DataObservableMixin(popup_renderer_1.PopupMixin(clusters_1.ClusterMixin(trace_1.TraceMixin(extend_bounds_1.ExtendBoundsMixin(div_markers_renderes_1.DivMarkersMixin(marker_renderer_1.MarkersMixin(styles_1.StyleMixin(map_items_1.MapItemsMixin(ActionsCluster)))))))));
