// ht-maps/dist/lib/entities/actions-data-polyline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var data_observable_1 = require("../mixins/data-observable");
var trace_1 = require("../mixins/trace");
var polyline_renderer_1 = require("../mixins/polyline-renderer");
var styles_1 = require("../mixins/styles");
var extend_bounds_1 = require("../mixins/extend-bounds");
var marker_renderer_1 = require("../mixins/marker-renderer");
var animations_entities_1 = require("../mixins/animations-entities");
var map_items_1 = require("../mixins/map-items");
var ActionsDataPolyline = (function () {
    function ActionsDataPolyline(mapInstance) {
        this.name = "Actions data polyline";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case 'google': {
                        return {
                            default: {},
                            popup: {
                                pixelOffset: new google.maps.Size(0, -5),
                            }
                        };
                    }
                    case 'leaflet': {
                        return {
                            default: {
                                fillColor: '#9155f4',
                                color: '#9155f4',
                                weight: 5
                                // iconAnchor: [12, 12]
                                // iconSize: [35, 35],
                                // className: 'current-action-marker',
                                // iconAnchor: point(15, 43)
                                // iconAnchor: [15, 43]
                            },
                            popup: {
                                // offset: point(0, -35),
                                offset: [0, -5],
                                closeButton: false
                            }
                        };
                    }
                }
            }
        };
        this.mapInstance = mapInstance;
    }
    ActionsDataPolyline.prototype.getEncodedPositionTime = function (data) {
        return data.time_aware_polyline;
    };
    //todo remove this, use getTimeAwarePolyline
    ActionsDataPolyline.prototype.getEncodedPath = function (data) {
        return data.encoded_polyline;
    };
    ActionsDataPolyline.prototype.getPosition = function (action) {
        var position = action.user ?
            action.user.last_location ? action.user.last_location.geojson.coordinates : null
            : null;
        return position ? { lat: position[1], lng: position[0] } : null;
    };
    ActionsDataPolyline.prototype.getTimeAwarePolyline = function (data) {
        return data.time_aware_polyline;
    };
    return ActionsDataPolyline;
}());
exports.ActionsDataPolyline = ActionsDataPolyline;
// export interface IActionsPolylineBase {
//   mapInstance: MapInstance
//   // getStyle: (styleType?) => object;
// }
//
// export function ActionsPolylineMixin<TBase extends Constructor<IActionsPolylineBase>>(Base: TBase) {
//   return class extends Base {
//     start;
//
//
//
//
//
//   };
// }
exports.ActionsPolylineTrace = data_observable_1.DataObservableMixin(animations_entities_1.AnimationsEntitiesMixin(trace_1.TraceMixin(extend_bounds_1.ExtendBoundsMixin(polyline_renderer_1.PolylinesMixin(marker_renderer_1.MarkersMixin(styles_1.StyleMixin(map_items_1.MapItemsMixin(ActionsDataPolyline))))))));
// export const ActionsPolylineTrace = DataObservableMixin(AnimationMixin(TraceMixin(PolylinesMixin(StyleMixin(ActionsDataPolyline)))))
