// ht-maps/dist/lib/entities/action-user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ht_data_1 = require("ht-data");
var div_markers_renderes_1 = require("../mixins/div-markers-renderes");
var trace_1 = require("../mixins/trace");
var marker_renderer_1 = require("../mixins/marker-renderer");
var styles_1 = require("../mixins/styles");
var data_observable_1 = require("../mixins/data-observable");
var extend_bounds_1 = require("../mixins/extend-bounds");
var animations_entities_1 = require("../mixins/animations-entities");
var map_items_1 = require("../mixins/map-items");
var ActionUser = (function () {
    function ActionUser(mapInstance) {
        this.mapInstance = mapInstance;
        this.name = "action user";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case "google":
                        {
                            return {
                                default: {
                                    flat: true,
                                    anchor: RichMarkerPosition.BOTTOM_CENTER
                                }
                            };
                        }
                        ;
                    case "leaflet": {
                        return {
                            default: {
                                zIndexOffset: 10,
                                iconSize: [60, 60],
                                className: "action-marker"
                                // iconAnchor: point(15, 43)
                                // iconAnchor: [0, 0]
                            }
                        };
                    }
                }
            }
        };
    }
    ActionUser.prototype.getPosition = function (data) {
        return ht_data_1.htUser(data.user).getPosition();
    };
    ;
    ActionUser.prototype.trackBy = function (action) {
        return action.user.id;
    };
    ActionUser.prototype.trackAnimationBy = function (action) {
        return action.id;
    };
    ActionUser.prototype.getDivContent = function (data) {
        var content = "\n    <div style=\"border-radius: 50%; height: 60px; width: 60px; background: rgba(95,143,213,0.67) \">\n  <div style=\"height: 30px; width: 30px; background-image: url('" + data.user.photo + "'); background-repeat: no-repeat;\n  background-size: cover;     top: 15px;\n    position: relative;\n    left: 15px;\n    border-radius: 50%;\"></div>\n</div> \n    ";
        return content;
    };
    ;
    return ActionUser;
}());
exports.ActionUser = ActionUser;
exports.ActionUserTrace = data_observable_1.DataObservableMixin(extend_bounds_1.ExtendBoundsMixin(animations_entities_1.AnimationsEntitiesMixin(div_markers_renderes_1.DivMarkersMixin(trace_1.TraceMixin(marker_renderer_1.MarkersMixin(styles_1.StyleMixin(map_items_1.MapItemsMixin(ActionUser))))))));
