// ht-maps/dist/lib/entities/actions-heatmap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var heatmap_1 = require("../mixins/heatmap");
var data_observable_1 = require("../mixins/data-observable");
var styles_1 = require("../mixins/styles");
var extend_bounds_1 = require("../mixins/extend-bounds");
var map_items_1 = require("../mixins/map-items");
var ActionsHeatmap = (function () {
    function ActionsHeatmap(mapInstance) {
        this.mapInstance = mapInstance;
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case "google": {
                        return {
                            default: {}
                        };
                    }
                    case "leaflet": {
                        return {
                            default: {
                                radius: 7,
                                minOpacity: 0.6
                            }
                        };
                    }
                }
            }
        };
    }
    ActionsHeatmap.prototype.getPosition = function (item) {
        return {
            lat: item.completed_place__location[1],
            lng: item.completed_place__location[0],
        };
    };
    ;
    return ActionsHeatmap;
}());
exports.ActionsHeatmap = ActionsHeatmap;
;
exports.ActionsHeatmapTrace = extend_bounds_1.ExtendBoundsMixin(data_observable_1.DataObservableMixin(extend_bounds_1.ExtendBoundsMixin(heatmap_1.HeatmapMixin(styles_1.StyleMixin(map_items_1.MapItemsMixin(ActionsHeatmap))))));
