// ht-maps/dist/lib/entities/action-markers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ht_data_1 = require("ht-data");
var marker_renderer_1 = require("../mixins/marker-renderer");
var styles_1 = require("../mixins/styles");
var trace_1 = require("../mixins/trace");
var extend_bounds_1 = require("../mixins/extend-bounds");
var div_markers_renderes_1 = require("../mixins/div-markers-renderes");
var ht_utility_1 = require("ht-utility");
var map_items_1 = require("../mixins/map-items");
var ActionMarkers = (function () {
    function ActionMarkers(mapInstance) {
        this.mapInstance = mapInstance;
        this.name = "Action";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case 'google': {
                        return {
                            default: {
                                flat: true,
                                anchor: RichMarkerPosition.MIDDLE,
                                zIndex: 1
                            },
                            popup: {
                                pixelOffset: new google.maps.Size(0, -5),
                            }
                        };
                    }
                    case 'leaflet': {
                        return {
                            default: {
                                iconAnchor: [12, 12]
                                // iconSize: [35, 35],
                                // className: 'current-action-marker',
                                // iconAnchor: point(15, 43)
                                // iconAnchor: [15, 43]
                            },
                            popup: {
                                // offset: point(0, -35),
                                offset: [0, -5],
                                closeButton: false
                            }
                        };
                    }
                }
            }
        };
    }
    ActionMarkers.prototype.getPosition = function (data) {
        var posObj = ht_data_1.htAction(data).getPositionsObject();
        return posObj ? posObj.position : null;
    };
    ActionMarkers.prototype.getDivContent = function (action) {
        var icon = "<div class=\"action-marker flex-row\">\n<span style=\"margin: auto\">" + ht_utility_1.NameCase(action.type[0]) + "</span>\n</div>";
        return icon;
    };
    return ActionMarkers;
}());
exports.ActionMarkers = ActionMarkers;
exports.ActionMarkersTrace = extend_bounds_1.ExtendBoundsMixin(trace_1.TraceMixin(div_markers_renderes_1.DivMarkersMixin(marker_renderer_1.MarkersMixin(styles_1.StyleMixin(map_items_1.MapItemsMixin(ActionMarkers))))));
