// ht-maps/dist/lib/entities/current-user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ht_data_1 = require("ht-data");
var user_div_factory_1 = require("../helpers/user-div-factory");
var map_items_factory_1 = require("../base/map-items-factory");
var single_item_1 = require("../mixins/single-item");
var div_markers_renderes_1 = require("../mixins/div-markers-renderes");
var trace_1 = require("../mixins/trace");
var marker_renderer_1 = require("../mixins/marker-renderer");
var styles_1 = require("../mixins/styles");
var animation_renderer_1 = require("../mixins/animation-renderer");
var map_items_1 = require("../mixins/map-items");
exports.currentUserConfig = {
    renderConfig: {
        getPosition: function (data) {
            return ht_data_1.htUser(data).getPosition();
        },
        getDivContent: function (data) {
            return user_div_factory_1.userDivFactory(data);
        }
    },
    styleFunct: {
        get: function (type) {
            switch (type) {
                case "google":
                    {
                        return {
                            default: {
                                flat: true,
                                anchor: RichMarkerPosition.BOTTOM_CENTER
                            }
                        };
                    }
                    ;
                case "leaflet": {
                    return {
                        default: {
                            // iconAnchor: point(15, 43)
                            iconAnchor: [15, 43]
                        }
                    };
                }
            }
        }
    },
    typeConfig: {
        isSingleItem: true,
        isDiv: true,
        hasDataObservable: false
    },
    name: "Current user"
};
exports.currentUserTrace = function () {
    return map_items_factory_1.itemsFactory(exports.currentUserConfig);
};
var CurrentUser = (function () {
    function CurrentUser(mapInstance) {
        this.mapInstance = mapInstance;
        this.name = "Current user";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case "google":
                        {
                            return {
                                default: {
                                    flat: true,
                                    anchor: RichMarkerPosition.BOTTOM_CENTER
                                }
                            };
                        }
                        ;
                    case "leaflet": {
                        return {
                            default: {
                                // iconAnchor: point(15, 43)
                                iconAnchor: [15, 43]
                            }
                        };
                    }
                }
            }
        };
    }
    CurrentUser.prototype.getPosition = function (data) {
        return ht_data_1.htUser(data).getPosition();
    };
    ;
    CurrentUser.prototype.getDivContent = function (data) {
        return user_div_factory_1.userDivFactory(data);
    };
    return CurrentUser;
}());
exports.CurrentUser = CurrentUser;
exports.CurrentUserTrace = animation_renderer_1.AnimationMixin(single_item_1.SingleItemMixin(div_markers_renderes_1.DivMarkersMixin(trace_1.TraceMixin(marker_renderer_1.MarkersMixin(styles_1.StyleMixin(map_items_1.MapItemsMixin(CurrentUser)))))));
// export class CurrentUser {
//   name = "Current user";
//   styleFunct: StyleFunct = {
//     google: {
//       default: {
//         flat: true,
//         anchor: RichMarkerPosition.BOTTOM_CENTER,
//       }
//     },
//     leaflet: {
//       default: {
//
//       }
//     }
//   };
//
//   getPosition(data): HtPosition {
//     return htUser(data).getPosition()
//   };
//
//   getDivContent(data) {
//     return userDivFactory(data)
//
//   }
// };
//
// export const CurrentUserTrace = mapItemsFactory(CurrentUser, {
//   isSingleItem: true,
//   isDiv: true,
//   hasDataObservable: false
// });
