// ht-maps/dist/lib/entities/destination-marker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var extend_bounds_1 = require("../mixins/extend-bounds");
var trace_1 = require("../mixins/trace");
var marker_renderer_1 = require("../mixins/marker-renderer");
var styles_1 = require("../mixins/styles");
var data_observable_1 = require("../mixins/data-observable");
var div_markers_renderes_1 = require("../mixins/div-markers-renderes");
var map_items_1 = require("../mixins/map-items");
var DestinationMap = (function () {
    function DestinationMap(mapInstance) {
        this.name = "Destination";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case 'google': {
                        return {
                            default: {},
                            popup: {
                                pixelOffset: new google.maps.Size(0, -5),
                            }
                        };
                    }
                    case 'leaflet': {
                        return {
                            default: {
                                zIndexOffset: 10,
                                iconSize: [30, 30],
                                className: "destination-marker"
                            }
                        };
                    }
                }
            }
        };
        this.mapInstance = mapInstance;
    }
    DestinationMap.prototype.getPosition = function (data) {
        var place = data.completed_place || data.expected_place;
        if (place && place.location) {
            return { lat: place.location.coordinates[1], lng: place.location.coordinates[0] };
        }
        else {
            return null;
        }
    };
    DestinationMap.prototype.getDivContent = function () {
        return "<div style=\"width: 30px; height: 30px; background: #00C94B; border-radius: 50%\"></div>";
    };
    return DestinationMap;
}());
exports.DestinationMap = DestinationMap;
;
exports.DestinationMarker = data_observable_1.DataObservableMixin(extend_bounds_1.ExtendBoundsMixin(trace_1.TraceMixin(div_markers_renderes_1.DivMarkersMixin(marker_renderer_1.MarkersMixin(styles_1.StyleMixin(map_items_1.MapItemsMixin(DestinationMap)))))));
