// ht-maps/dist/lib/entities/segment-polylines.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var segment_polyline_styles_1 = require("../styles/segment-polyline-styles");
var extend_bounds_1 = require("../mixins/extend-bounds");
var marker_renderer_1 = require("../mixins/marker-renderer");
var styles_1 = require("../mixins/styles");
var trace_1 = require("../mixins/trace");
var polyline_renderer_1 = require("../mixins/polyline-renderer");
var map_items_1 = require("../mixins/map-items");
var SegmentPolylines = (function () {
    function SegmentPolylines(mapInstance) {
        this.mapInstance = mapInstance;
        this.styleFunct = segment_polyline_styles_1.segmentPolylineStyles;
        this.name = "segment polyline";
    }
    SegmentPolylines.prototype.getEncodedPath = function (data) {
        return data.encoded_polyline;
    };
    ;
    SegmentPolylines.prototype.getEncodedPositionTime = function (data) {
        return data.time_aware_polyline;
    };
    SegmentPolylines.prototype.getPosition = function () {
        return { lat: 0, lng: 0 };
    };
    return SegmentPolylines;
}());
exports.SegmentPolylines = SegmentPolylines;
exports.SegmentPolylinesTrace = trace_1.TraceMixin(extend_bounds_1.ExtendBoundsMixin(polyline_renderer_1.PolylinesMixin(marker_renderer_1.MarkersMixin(styles_1.StyleMixin(map_items_1.MapItemsMixin(SegmentPolylines))))));
// export class SegmentPolylines {
//   styleFunct = segmentPolylineStyles;
//
//   getEncodedPath(data) {
//     return data.encoded_polyline;
//   };
//
//   getEncodedPositionTime(data: ISegment) {
//     return data.time_aware_polyline
//   }
// }
//
// export const SegmentPolylinesTrace = mapItemsFactory(SegmentPolylines, {
//   isPolyline: true,
//   hasDataObservable: false
// });
// export const SegmentPolylinesTrace = _.compose(
//   PolylinesMixin,
//   MarkersMixin,
//   StyleMixin,
//   TraceMixin,
// )(SegmentPolylines);
