// ht-maps/dist/lib/entities/start-marker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var extend_bounds_1 = require("../mixins/extend-bounds");
var trace_1 = require("../mixins/trace");
var marker_renderer_1 = require("../mixins/marker-renderer");
var styles_1 = require("../mixins/styles");
var data_observable_1 = require("../mixins/data-observable");
var div_markers_renderes_1 = require("../mixins/div-markers-renderes");
var map_items_1 = require("../mixins/map-items");
var StartMarker = (function () {
    function StartMarker(mapInstance) {
        this.name = "Start marker";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case 'google': {
                        return {
                            default: {},
                            popup: {
                                pixelOffset: new google.maps.Size(0, -5),
                            }
                        };
                    }
                    case 'leaflet': {
                        return {
                            default: {
                                zIndexOffset: 10,
                                iconSize: [20, 20],
                                className: "destination-marker"
                            }
                        };
                    }
                }
            }
        };
        this.mapInstance = mapInstance;
    }
    StartMarker.prototype.trace = function (mapItem) {
    };
    StartMarker.prototype.getPosition = function (data) {
        var place = data.started_place;
        if (place && place.location) {
            return { lat: place.location.coordinates[1], lng: place.location.coordinates[0] };
        }
        else {
            return null;
        }
    };
    StartMarker.prototype.getDivContent = function () {
        return "<div style=\"width: 20px; height: 20px; background: #9013FE; border-radius: 50%\"></div>";
    };
    return StartMarker;
}());
exports.StartMarker = StartMarker;
;
exports.StartMarkerTrace = extend_bounds_1.ExtendBoundsMixin(data_observable_1.DataObservableMixin(trace_1.TraceMixin(div_markers_renderes_1.DivMarkersMixin(marker_renderer_1.MarkersMixin(styles_1.StyleMixin(map_items_1.MapItemsMixin(StartMarker)))))));
