// ht-maps/dist/lib/entities/stops-heatmap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var heatmap_1 = require("../mixins/heatmap");
var data_observable_1 = require("../mixins/data-observable");
var styles_1 = require("../mixins/styles");
var extend_bounds_1 = require("../mixins/extend-bounds");
var map_items_1 = require("../mixins/map-items");
var StopsHeatmap = (function () {
    function StopsHeatmap(mapInstance) {
        this.name = "stop heatmap";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case "google": {
                        return {
                            default: {}
                        };
                    }
                    case "leaflet": {
                        return {
                            default: {
                                radius: 7,
                                minOpacity: 0.6
                            }
                        };
                    }
                }
            }
        };
        this.mapInstance = mapInstance;
    }
    StopsHeatmap.prototype.getPosition = function (item) {
        return {
            lat: item.place__location[1],
            lng: item.place__location[0],
            weight: item.intensity
        };
    };
    return StopsHeatmap;
}());
exports.StopsHeatmap = StopsHeatmap;
;
exports.StopsHeatmapTrace = extend_bounds_1.ExtendBoundsMixin(data_observable_1.DataObservableMixin(heatmap_1.HeatmapMixin(styles_1.StyleMixin(map_items_1.MapItemsMixin(StopsHeatmap)))));
