// ht-maps/dist/lib/entities/stop-markers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var stop_styles_1 = require("../styles/stop-styles");
var ht_utility_1 = require("ht-utility");
var map_items_factory_1 = require("../base/map-items-factory");
var popup_renderer_1 = require("../mixins/popup-renderer");
var trace_1 = require("../mixins/trace");
var extend_bounds_1 = require("../mixins/extend-bounds");
var circle_renderer_1 = require("../mixins/circle-renderer");
var marker_renderer_1 = require("../mixins/marker-renderer");
var styles_1 = require("../mixins/styles");
var map_items_1 = require("../mixins/map-items");
exports.stopMarkersConfig = {
    renderConfig: {
        getPosition: function (data) {
            if (data.location && data.location.geojson) {
                var lat = data.location.geojson.coordinates[1];
                var lng = data.location.geojson.coordinates[0];
                return { lat: lat, lng: lng };
            }
            else {
                return null;
            }
        },
        getInfoContent: function (data) {
            var durationString = null;
            if (data.started_at && data.ended_at) {
                var durationMin = new Date(new Date(data.ended_at).getTime() -
                    new Date(data.started_at).getTime()).getTime() /
                    (1000 * 60);
                durationString = ht_utility_1.HMString(durationMin);
            }
            var start = ht_utility_1.TimeString(data.started_at);
            var end = ht_utility_1.TimeString(data.ended_at);
            var startDate = ht_utility_1.DateString(data.started_at);
            var startDateShort = ht_utility_1.DateString(data.started_at, "short");
            var endDate = ht_utility_1.DateString(data.ended_at);
            var sameDate = startDate == endDate;
            function htShow(item) {
                return "display: " + (item ? "flex" : "none");
            }
            return "<div class=\"flex-column\">\n<strong class=\"text-muted text-center\" style=\"padding-bottom: 0; color: " + ht_utility_1.Color.stop + "\">Stop</strong>\n<div style=\"display: " + (start ? "display" : "none") + "\" class=\"flex-row space-between\">\n    <div>" + (start ||
                "--:--") + "</div><div>&nbsp; to &nbsp;</div><div class=\"text-right\">" + (end ||
                "--:--") + "</div>\n</div>\n<div style=\"" + htShow(startDate || endDate) + "\" class=\"" + (sameDate || !endDate ? "space-around" : "space-between") + " text-2 text-muted flex-row\">\n    <div>" + startDateShort + "</div><div style=\"display: " + (sameDate || !endDate ? "none" : "block") + "\">" + endDate + "</div>\n</div>\n<div style=\"display: " + (durationString ? "block" : "none") + "\" class=\"text-3 text-center\">" + durationString + "</div>\n</div>";
        }
    },
    styleFunct: stop_styles_1.stopStyles,
    typeConfig: {
        hasPopup: true,
        isCircle: true,
        hasDataObservable: false
    },
    name: "stop"
};
exports.stopMarkersTrace = function () {
    return map_items_factory_1.itemsFactory(exports.stopMarkersConfig);
};
var StopMarkers = (function () {
    function StopMarkers(mapInstance) {
        this.mapInstance = mapInstance;
        this.styleFunct = stop_styles_1.stopStyles;
    }
    StopMarkers.prototype.getPosition = function (data) {
        if (data.location && data.location.geojson) {
            var lat = data.location.geojson.coordinates[1];
            var lng = data.location.geojson.coordinates[0];
            return { lat: lat, lng: lng };
        }
        else {
            return null;
        }
    };
    ;
    StopMarkers.prototype.getInfoContent = function (data) {
        var durationString = null;
        if (data.started_at && data.ended_at) {
            var durationMin = new Date(new Date(data.ended_at).getTime() -
                new Date(data.started_at).getTime()).getTime() /
                (1000 * 60);
            durationString = ht_utility_1.HMString(durationMin);
        }
        var start = ht_utility_1.TimeString(data.started_at);
        var end = ht_utility_1.TimeString(data.ended_at);
        var startDate = ht_utility_1.DateString(data.started_at);
        var startDateShort = ht_utility_1.DateString(data.started_at, "short");
        var endDate = ht_utility_1.DateString(data.ended_at);
        var sameDate = startDate == endDate;
        function htShow(item) {
            return "display: " + (item ? "flex" : "none");
        }
        return "<div class=\"flex-column\">\n<strong class=\"text-muted text-center\" style=\"padding-bottom: 0; color: " + ht_utility_1.Color.stop + "\">Stop</strong>\n<div style=\"display: " + (start ? "display" : "none") + "\" class=\"flex-row space-between\">\n    <div>" + (start ||
            "--:--") + "</div><div>&nbsp; to &nbsp;</div><div class=\"text-right\">" + (end ||
            "--:--") + "</div>\n</div>\n<div style=\"" + htShow(startDate || endDate) + "\" class=\"" + (sameDate || !endDate ? "space-around" : "space-between") + " text-2 text-muted flex-row\">\n    <div>" + startDateShort + "</div><div style=\"display: " + (sameDate || !endDate ? "none" : "block") + "\">" + endDate + "</div>\n</div>\n<div style=\"display: " + (durationString ? "block" : "none") + "\" class=\"text-3 text-center\">" + durationString + "</div>\n</div>";
    };
    return StopMarkers;
}());
exports.StopMarkers = StopMarkers;
exports.StopMarkersTrace = popup_renderer_1.PopupMixin(trace_1.TraceMixin(extend_bounds_1.ExtendBoundsMixin(circle_renderer_1.CircleMixin(marker_renderer_1.MarkersMixin(styles_1.StyleMixin(map_items_1.MapItemsMixin(StopMarkers)))))));
// export abstract class StopMarkers {
//   name = "Stop";
//   styleFunct = stopStyles;
//
//   htShow(item) {
//     return `display: ${item ? 'flex' : 'none'}`
//   };
//
//   getPosition(data) {
//     if(data.location && data.location.geojson) {
//       let lat = data.location.geojson.coordinates[1];
//       let lng = data.location.geojson.coordinates[0];
//       return {lat, lng}
//     } else {
//       return null;
//     }
//
//   };
//
//   getInfoContent(data: ISegment) {
//     function htShow(item) {
//       return `display: ${item ? 'flex' : 'none'}`
//     };
//     console.log(this.htShow, "htShow");
//     let durationString = null;
//     if(data.started_at && data.ended_at) {
//       let durationMin = (new Date(new Date(data.ended_at).getTime() - new Date(data.started_at).getTime()).getTime()) / (1000 * 60);
//       durationString = HMString(durationMin)
//     }
//
//     let start = TimeString(data.started_at);
//     let end = TimeString(data.ended_at);
//     let startDate = DateString(data.started_at);
//     let startDateShort = DateString(data.started_at, 'short');
//     let endDate = DateString(data.ended_at);
//     let sameDate = startDate == endDate;
//     return `<div class="flex-column">
// <strong class="text-muted text-center" style="padding-bottom: 0; color: ${Color.stop}">Stop</strong>
// <div style="display: ${start ? 'display' : 'none'}" class="flex-row space-between">
//     <div>${start || '--:--'}</div><div>&nbsp; to &nbsp;</div><div class="text-right">${end || '--:--'}</div>
// </div>
// <div style="${htShow(startDate || endDate)}" class="${sameDate || !endDate ? 'space-around' : 'space-between'} text-2 text-muted flex-row">
//     <div>${startDateShort}</div><div style="display: ${sameDate || !endDate ? 'none' : 'block'}">${endDate}</div>
// </div>
// <div style="display: ${durationString ? 'block' : 'none'}" class="text-3 text-center">${durationString}</div>
// </div>`
//   }
// }
//
// export const StopMarkersTrace = mapItemsFactory(StopMarkers, {
//   hasPopup: true,
//   isCircle: true,
//   hasDataObservable: false
// });
