// ht-maps/dist/lib/entities/users-cluster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ht_data_1 = require("ht-data");
var user_div_factory_1 = require("../helpers/user-div-factory");
var extend_bounds_1 = require("../mixins/extend-bounds");
var marker_renderer_1 = require("../mixins/marker-renderer");
var styles_1 = require("../mixins/styles");
var data_observable_1 = require("../mixins/data-observable");
var popup_renderer_1 = require("../mixins/popup-renderer");
var div_markers_renderes_1 = require("../mixins/div-markers-renderes");
var trace_1 = require("../mixins/trace");
var clusters_1 = require("../mixins/clusters");
var map_items_1 = require("../mixins/map-items");
var UsersCluster = (function () {
    function UsersCluster(mapInstance) {
        this.mapInstance = mapInstance;
        this.name = "users cluster";
        this.styleFunct = {
            get: function (type) {
                switch (type) {
                    case 'google': {
                        return {
                            default: {
                                flat: true,
                                anchor: RichMarkerPosition.BOTTOM_CENTER,
                                zIndex: 1
                            },
                            popup: {
                                disableAutoPan: true,
                                pixelOffset: new google.maps.Size(0, -35)
                            }
                        };
                    }
                    case 'leaflet': {
                        return {
                            default: {
                                // iconAnchor: point(15, 43)
                                iconAnchor: [15, 43]
                            },
                            popup: {
                                // offset: point(0, -35),
                                offset: [0, -35],
                                closeButton: false
                            }
                        };
                    }
                }
            }
        };
    }
    UsersCluster.prototype.getPosition = function (data) {
        return ht_data_1.htUser(data).getPosition();
    };
    ;
    UsersCluster.prototype.getDivContent = function (data) {
        return user_div_factory_1.userDivFactory(data);
    };
    ;
    UsersCluster.prototype.getInfoContent = function (data) {
        var string = "<div>\n<strong>" + data.name + "</strong>\n<div>" + data.display.status_text + "</div>\n<div>" + data.display.sub_status_text + "</div>\n</div>";
        return string;
    };
    UsersCluster.prototype.traceEffect = function () {
    };
    return UsersCluster;
}());
exports.UsersCluster = UsersCluster;
exports.UsersClusterTrace = data_observable_1.DataObservableMixin(popup_renderer_1.PopupMixin(clusters_1.ClusterMixin(trace_1.TraceMixin(extend_bounds_1.ExtendBoundsMixin(div_markers_renderes_1.DivMarkersMixin(marker_renderer_1.MarkersMixin(styles_1.StyleMixin(map_items_1.MapItemsMixin(UsersCluster)))))))));
// export class UsersClusters {
//   name = "Cluster user";
//   styleFunct: StyleFunct = {
//     google: {
//       default: {
//         flat: true,
//         anchor: RichMarkerPosition.BOTTOM_CENTER,
//         zIndex: 1
//       },
//       popup: {
//         disableAutoPan: true,
//         pixelOffset: new google.maps.Size(0, -35)
//       }
//     },
//     leaflet: {
//       default: {
//
//       }
//     }
//   };
//
//   getPosition(data): HtPosition {
//     return htUser(data).getPosition()
//   };
//
//   getDivContent(data) {
//     return userDivFactory(data)
//   };
//
//   getInfoContent(data) {
//     let string = `<div>
// <strong>${data.name}</strong>
// <div>${data.display.status_text}</div>
// <div>${data.display.sub_status_text}</div>
// </div>`;
//     return string
//   }
// }
//
// export const UsersClustersTrace = mapItemsFactory(UsersClusters, {
//   hasPopup: true,
//   isCluster: true,
//   isDiv: true,
//   hasDataObservable: true
// });
